<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:st1="urn:schemas-microsoft-com:office:smarttags"
xmlns="http://www.w3.org/TR/REC-html40">

<head>
<meta http-equiv=Content-Type content="text/html; charset=windows-1252">
<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 11">
<meta name=Originator content="Microsoft Word 11">
<link rel=File-List href="gep-extension_files/filelist.xml">
<title>ECJ</title>
<o:SmartTagType namespaceuri="urn:schemas-microsoft-com:office:smarttags"
 name="PlaceType"/>
<o:SmartTagType namespaceuri="urn:schemas-microsoft-com:office:smarttags"
 name="PlaceName"/>
<o:SmartTagType namespaceuri="urn:schemas-microsoft-com:office:smarttags"
 name="City"/>
<o:SmartTagType namespaceuri="urn:schemas-microsoft-com:office:smarttags"
 name="State"/>
<o:SmartTagType namespaceuri="urn:schemas-microsoft-com:office:smarttags"
 name="PersonName"/>
<o:SmartTagType namespaceuri="urn:schemas-microsoft-com:office:smarttags"
 name="country-region"/>
<o:SmartTagType namespaceuri="urn:schemas-microsoft-com:office:smarttags"
 name="place"/>
<!--[if gte mso 9]><xml>
 <o:DocumentProperties>
  <o:Author>Bob Orchard</o:Author>
  <o:LastAuthor> </o:LastAuthor>
  <o:Revision>21</o:Revision>
  <o:TotalTime>313</o:TotalTime>
  <o:Created>2008-08-05T13:17:00Z</o:Created>
  <o:LastSaved>2008-12-09T20:26:00Z</o:LastSaved>
  <o:Pages>1</o:Pages>
  <o:Words>13648</o:Words>
  <o:Characters>77797</o:Characters>
  <o:Company>National Research Council</o:Company>
  <o:Lines>648</o:Lines>
  <o:Paragraphs>182</o:Paragraphs>
  <o:CharactersWithSpaces>91263</o:CharactersWithSpaces>
  <o:Version>11.9999</o:Version>
 </o:DocumentProperties>
</xml><![endif]--><!--[if gte mso 9]><xml>
 <w:WordDocument>
  <w:GrammarState>Clean</w:GrammarState>
  <w:ValidateAgainstSchemas/>
  <w:SaveIfXMLInvalid>false</w:SaveIfXMLInvalid>
  <w:IgnoreMixedContent>false</w:IgnoreMixedContent>
  <w:AlwaysShowPlaceholderText>false</w:AlwaysShowPlaceholderText>
  <w:BrowserLevel>MicrosoftInternetExplorer4</w:BrowserLevel>
 </w:WordDocument>
</xml><![endif]--><!--[if gte mso 9]><xml>
 <w:LatentStyles DefLockedState="false" LatentStyleCount="156">
 </w:LatentStyles>
</xml><![endif]--><!--[if !mso]><object
 classid="clsid:38481807-CA0E-42D2-BF39-B33AF135CC4D" id=ieooui></object>
<style>
st1\:*{behavior:url(#ieooui) }
</style>
<![endif]-->
<style>
<!--
 /* Font Definitions */
 @font-face
	{font-family:Helvetica;
	panose-1:2 11 6 4 2 2 2 2 2 4;
	mso-font-charset:0;
	mso-generic-font-family:swiss;
	mso-font-pitch:variable;
	mso-font-signature:536902279 -2147483648 8 0 511 0;}
 /* Style Definitions */
 p.MsoNormal, li.MsoNormal, div.MsoNormal
	{mso-style-parent:"";
	margin:0in;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";}
h1
	{mso-style-next:Normal;
	margin:0in;
	margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	page-break-after:avoid;
	mso-outline-level:1;
	font-size:12.0pt;
	font-family:"Times New Roman";
	mso-font-kerning:0pt;
	font-weight:bold;}
a:link, span.MsoHyperlink
	{color:blue;
	text-decoration:underline;
	text-underline:single;}
a:visited, span.MsoHyperlinkFollowed
	{color:purple;
	text-decoration:underline;
	text-underline:single;}
p
	{mso-margin-top-alt:auto;
	margin-right:0in;
	mso-margin-bottom-alt:auto;
	margin-left:0in;
	mso-pagination:widow-orphan;
	font-size:12.0pt;
	font-family:"Times New Roman";
	mso-fareast-font-family:"Times New Roman";}
code
	{font-family:"Courier New";
	mso-ascii-font-family:"Courier New";
	mso-fareast-font-family:"Times New Roman";
	mso-hansi-font-family:"Courier New";
	mso-bidi-font-family:"Courier New";}
p.1Level, li.1Level, div.1Level
	{mso-style-name:1Level;
	mso-style-update:auto;
	margin-top:0in;
	margin-right:0in;
	margin-bottom:0in;
	margin-left:.25in;
	margin-bottom:.0001pt;
	text-indent:-.25in;
	mso-pagination:none;
	tab-stops:89.4pt;
	mso-layout-grid-align:none;
	text-autospace:none;
	font-size:14.0pt;
	mso-bidi-font-size:12.0pt;
	font-family:Helvetica;
	mso-fareast-font-family:"Times New Roman";
	mso-bidi-font-family:"Times New Roman";
	font-weight:bold;
	mso-bidi-font-weight:normal;
	font-style:italic;
	mso-bidi-font-style:normal;}
span.GramE
	{mso-style-name:"";
	mso-gram-e:yes;}
@page Section1
	{size:8.5in 11.0in;
	margin:1.0in 1.25in 1.0in 1.25in;
	mso-header-margin:.5in;
	mso-footer-margin:.5in;
	mso-paper-source:0;}
div.Section1
	{page:Section1;}
-->
</style>
<!--[if gte mso 10]>
<style>
 /* Style Definitions */
 table.MsoNormalTable
	{mso-style-name:"Table Normal";
	mso-tstyle-rowband-size:0;
	mso-tstyle-colband-size:0;
	mso-style-noshow:yes;
	mso-style-parent:"";
	mso-padding-alt:0in 5.4pt 0in 5.4pt;
	mso-para-margin:0in;
	mso-para-margin-bottom:.0001pt;
	mso-pagination:widow-orphan;
	font-size:10.0pt;
	font-family:"Times New Roman";
	mso-ansi-language:#0400;
	mso-fareast-language:#0400;
	mso-bidi-language:#0400;}
</style>
<![endif]--><!--[if gte mso 9]><xml>
 <o:shapedefaults v:ext="edit" spidmax="9218">
  <o:colormenu v:ext="edit" fillcolor="none"/>
 </o:shapedefaults></xml><![endif]--><!--[if gte mso 9]><xml>
 <o:shapelayout v:ext="edit">
  <o:idmap v:ext="edit" data="1"/>
 </o:shapelayout></xml><![endif]-->
</head>

<body lang=EN-US link=blue vlink=purple style='tab-interval:.5in'>

<div class=Section1>

<div>

<table class=MsoNormalTable border=0 cellpadding=0 width="100%"
 style='width:100.0%;mso-cellspacing:1.5pt;mso-padding-alt:0in 0in 0in 0in'>
 <tr style='mso-yfti-irow:0;mso-yfti-firstrow:yes;mso-yfti-lastrow:yes'>
  <td width="100%" style='width:100.0%;background:darkblue;padding:.75pt .75pt .75pt .75pt'>
  <p class=MsoNormal style='tab-stops:189.0pt'><b><span style='font-size:18.0pt;
  color:white'>ECJ</span></b> </p>
  </td>
 </tr>
</table>

<p style='margin-bottom:12.0pt'><b>An Evolutionary Computation and Genetic
Programming System</b></p>

<h1>Extension to support Gene Expression Programming (GEP)</h1>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p style='margin:0in;margin-bottom:.0001pt'>Bob Orchard
(bob.orchard@nrc-cnrc.gc.ca)</p>

<p class=MsoNormal>National Research Council of <st1:country-region w:st="on"><st1:place
 w:st="on">Canada</st1:place></st1:country-region></p>

<p class=MsoNormal>Institute for Information Technology</p>

<h1><o:p>&nbsp;</o:p></h1>

<h1><a href="#What_is_GEP">What is Gene Expression Programming?</a></h1>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<h1><a href="#How_to_use_GEP_in_ECJ">How Do I Use GEP in the ECJ System? <span
class=GramE>Some Simple Examples.</span></a></h1>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<h1><a href="#Some_examples">Some Other Examples</a></h1>

<h1><a href="#Function_Finding_Examples"><br>
</a></h1>

<div style='margin-left:30.0pt'>

<h1><a href="#Function_Finding_Examples">Function Finding Examples</a></h1>

</div>

<h1 style='margin-left:30.0pt'><a href="#Some_examples"><br>
Classification Examples</a></h1>

<div style='margin-left:30.0pt'>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

</div>

<h1 style='margin-left:30.0pt'><a href="#Time_Series_Examples">Time Series
Examples</a></h1>

<div style='margin-left:30.0pt'>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

</div>

<h1 style='margin-left:30.0pt'><a href="#Logical_Problem_Examples">Logical
Problem Examples</a></h1>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<h1><a href="#A_Detailed_Look_at_the_GEP_Parameter">A Detailed Look at the GEP
Parameter File (gep.param)</a></h1>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<h1><a href="#Some_of_the_Implementation_Details">Some of the Implementation
Details</a></h1>

<h1><o:p>&nbsp;</o:p></h1>

<div style='margin-left:30.0pt'>

<p class=MsoNormal><b><a href="#ModsToExistingECJcode">Changes to Existing ECJ
code</a><br>
<br>
<a href="#AddingNewFunctions">Adding New Math Functions and Fitness Functions</a><o:p></o:p></b></p>

</div>

<p class=MsoNormal style='margin-bottom:12.0pt'><o:p>&nbsp;</o:p></p>

<h1><a name="What_is_GEP">What is Gene Expression Programming</a>?</h1>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>There are a
number of topics that fall under the general category of Evolutionary
Computing. These include Genetic Algorithms (GA) and Genetic Programs (GP). <b>Gene
Expression Programming</b> (GEP) is a ‘new’ evolutionary algorithm (developed
by Cândida Ferreira) that evolves computer programs in a way similar to GP. But
the representation of the programs is more efficient and leads to less complex
implementation of the genetic operators. The individuals of gene expression
programming are encoded in linear chromosomes which are expressed or translated
into expression trees when evaluation is required. Thus, in GEP, the genotype
(the linear chromosomes) and the phenotype (the expression trees) are different
entities (both structurally and functionally).</span></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>As in nature, the
linear chromosomes consist of the genetic material that is passed on with
modification to the next generation. Therefore, in GEP, all the genetic
modifications take place in the chromosomes, and only the chromosomes are
transmitted in the process of reproduction. After reproduction the new
chromosomes are expressed forming the body or expression trees (ETs).</span></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>The ETs are
computer programs evolved to solve a particular problem and are selected
according to their capabilities in solving the problem at hand. With time,
populations of such computer programs discover new traits and become better
adapted to a particular selection environment (for instance, a set of
experimental results) and, hopefully, a good solution evolves.<o:p></o:p></span></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>This is variation
of Genetic Programming&nbsp;and not a ‘new’ type of evolutionary computing.
However, due to the genotype/phenotype representation and to the multigenic
organization of GEP chromosomes, this new algorithm generally leads to better
performance. It may also generate better solutions in many cases.<o:p></o:p></span></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>For all the
details of this new algorithm, see the seminal <a
href="http://www.gene-expression-programming.com/webpapers/gep.pdf"
title="http://www.gene-expression-programming.com/webpapers/gep.pdf">GEP paper</a>
by Ferreira (published in Complex Systems) where the algorithm is fully
described and applied to a set of problems, including symbolic regression, <span
class=GramE>boolean</span> concept learning, and cellular automata.<o:p></o:p></span></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>Now much of the
above is taken directly from the Wikipedia definition of GEP and as it says,
the details can be found in Ferreira’s paper and in much greater detail in her
excellent book, <i>Gene Expression programming: mathematical modeling by an
artificial intelligence.</i> <span class=GramE>2<sup>nd</sup> Edition.</span>
Springer-Verlag: 2006. <span class=GramE>ISBN 3-540-32796-7.</span> There is
also a commercial implementation of the algorithm by the company founded by
Ferreira, Gepsoft (<a href="http://www.gepsoft.com/">www.gepsoft.com</a>). The
product is called GeneXproTools. You might also look at the documentation
provided with the fully functioning (but limited) download of GeneXproTools to
gain much more insight and detail about gep, or just view the short tutorial (<a
href="http://www.gepsoft.com/tutorial002.htm">http://www.gepsoft.com/tutorial002.htm</a>).
<o:p></o:p></span></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>The intent of
this ECJ implementation of the gep algorithm is to provide a free, and
hopefully useful and accurate implementation of the software, with most of the
major functionality. And, since the source code is provided, researchers and
other interested parties can experiment with ways to enhance or improve some
aspect of this approach to evolutionary computing. </span></p>

<p class=MsoNormal>&nbsp;</p>

<h1><a name="How_to_use_GEP_in_ECJ">How Do I Use GEP in the ECJ System? </a><span
class=GramE><span style='mso-bookmark:How_to_use_GEP_in_ECJ'>Some Simple
Examples.</span></span><span style='mso-bookmark:How_to_use_GEP_in_ECJ'></span></h1>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal style='margin-bottom:12.0pt'>One doesn’t need to understand
the detailed implementation of gep in ECJ to use the system but it is necessary
to know about some things so that the appropriate parameters can be set and
your data can be processed.</p>

<div style='margin-left:60.0pt'>

<p class=MsoNormal><b><i>Please NOTE that the output from any examples in this
documentation may be different than the output you see when you execute the
problem</i></b><i>s</i>. The code is changing and this can cause changes to the
output even with the same input (especially if code changes affect the way or
order in which random numbers are allocated).</p>

</div>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>You need to supply 3 things to run an experiment: a <b><i>parameter
file</i></b> that defines how to do the evolution, a very <b><i>simple Java
class</i></b> (actually you don<st1:PersonName w:st="on">'</st1:PersonName>t
even need to do any programming at all ... more about that option later), and <b><i>some
data</i></b> (the data can be in a file or encoded in your simple Java class).
We’ll start by looking at a very simple problem, showing the simplest of
parameter files and the minimal Java program. </p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>The example is to determine a model (equation) that
satisfies the following set of data (stored in the file test1.txt). The
variables x and y are independent variables and z is the dependent variable, <span
class=GramE>so &nbsp;z</span> = f(x,y).</p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><span class=GramE><code><span style='font-size:10.0pt'>x</span></code></span><code><span
style='font-size:10.0pt'>,&nbsp;&nbsp;&nbsp;&nbsp; y,&nbsp;&nbsp;&nbsp; z<o:p></o:p></span></code></p>

<p class=MsoNormal><code><span style='font-size:10.0pt'>0.5,<span class=GramE>&nbsp;
1.0</span>,&nbsp; 1.25 <o:p></o:p></span></code></p>

<p class=MsoNormal><code><span style='font-size:10.0pt'>1.0,<span class=GramE>&nbsp;
2.0</span>,&nbsp; 3.0 <o:p></o:p></span></code></p>

<p class=MsoNormal><code><span style='font-size:10.0pt'>2.0,<span class=GramE>&nbsp;
4.0</span>,&nbsp; 8.0<o:p></o:p></span></code></p>

<p class=MsoNormal><code><span style='font-size:10.0pt'>3.0, 12.0, 21.0</span></code></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>You might quickly see that this is ‘perfect’ sample data for
the function z = x*x + y.&nbsp;Data files that are currently supported are the
so called CSV, comma separated value, text files. NOTE: gzip<st1:PersonName
w:st="on">'</st1:PersonName>d files can also be used directly. So if a file
ends in .gz or .gzip we will handle the de-compression of the file. For example
in this case we could have compressed test1.txt to create test1.txt.gz and
identified the gzip<st1:PersonName w:st="on">'</st1:PersonName>d file rather
than the .txt file in the parameter file (see below)<span class=GramE>.</span></p>

<p class=MsoNormal style='margin-bottom:12.0pt'>Generally one will use either a
comma, a space or a tab to be the field separator. The example above uses a
comma. The first row of the file must identify the independent and dependent
variable names with the last column being the dependent variable. You’ll see
later that for time series problems there is another file format that is used
with ‘raw’ time series data, but for most problems the training data, when in a
file, has this format. So here we have 3 columns that identify the 2
independent variables <i>x</i> and <i>y</i> plus the dependent variable <i>z</i>.
There are 4 rows of numeric data that we will use as <st1:PersonName w:st="on">'</st1:PersonName>training<st1:PersonName
w:st="on">'</st1:PersonName> data during the evolutionary process.&nbsp;</p>

<p class=MsoNormal>The parameter file that tells the gep system how to behave
as it generates models and proceeds from generation to generation is
test1.param. It has the following content:&nbsp;</p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><code><span style='font-size:10.0pt'># This parameter file
supports a simple test example to solve x^2 + y<span class=GramE>.</span></span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>#</code><br>
<code># The data (including variable names) is in the file test1.txt.</code><br>
<br>
<code># My parent<span class=GramE>:</span></code><br>
<code>parent.0 =&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; ../../../gep/gep.params</code><br>
<br>
<code># Problem</code><br>
<code>#===============================</code><br>
<code>eval.problem = ec.app.gep.test1.test1</code><br>
<br>
<code># run for 100 generations, quit prematurely if I find something ideal</code><br>
<code>generations =&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; 100</code><br>
<code>quit-on-run-complete =&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;true</code><br>
<br>
<code># ec.Subpopulation</code><br>
<code># ==============================</code><br>
<br>
<code># subpop size is 50 individuals</code><br>
<code>pop.subpop.0.size =&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; 50</code><br>
<br>
<code># ec.Species</code><br>
<code># ==============================</code><br>
<br>
<code>gep.species.numgenes = 2</code><br>
<code>gep.species.gene-headsize = 5</code><br>
<code>gep.species.gene-linking-function = +</code><br>
<br>
<code># Problem type must be one of: functionfinding, classification,
timeseries, logical</code><br>
<code># Set default to be <st1:PersonName w:st="on">'</st1:PersonName>unknown<st1:PersonName
w:st="on">'</st1:PersonName> so user is forced to specify in the problem params
file </code><br>
<code>gep.species.problemtype = functionfinding</code><br>
<br>
<code># specify the symbols ... functions and terminals (variables)</code><br>
<code>#</code><br>
<code># the terminal symbols (variable names) will be obtained from the data
file (1st row)</code><br>
<code>gep.species.symbolset.terminalfilename = ec/app/gep/test1/test1.txt</code><br>
<code># 4 functions + - * /&nbsp; are used in this problem</code><br>
<code>gep.species.symbolset.functionsize = 4</code><br>
<code>gep.species.symbolset.function.0 = Add</code><br>
<code>gep.species.symbolset.function.1 = Sub</code><br>
<code>gep.species.symbolset.function.2 = Mul</code><br>
<code>gep.species.symbolset.function.3 = Div</code><br>
<br>
<code>#&nbsp;Statistics</code><br>
<code># ==============================</code><br>
<br>
<code># output statistics to the file &quot;test1.stat&quot; in the directory</code><br>
<code># the run was started in</code><br>
<code>stat =&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; ec.gep.GEPSimpleStatistics</code><br>
<code>stat.file =&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;test1.stat</code></span></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>Let<st1:PersonName
w:st="on">'</st1:PersonName>s look at the parameters in some detail. First we
see that the file references a parent file (</span><span class=GramE><code><span
style='font-size:10.0pt'>..</span></code></span><code><span style='font-size:
10.0pt'>/../../gep/gep.params</span></code><span lang=EN style='mso-ansi-language:
EN'>). This is the parameter file that holds all of the default settings for a
gep problem. You can over-ride any of the settings in this file by including
new values in your own param file (in this case, test1.param). But your param
file must reference this file so that the settings that you are not concerned
about are properly defined. If you looked at the gep.params file you<st1:PersonName
w:st="on">'</st1:PersonName>d see the full set of parameters that can be specified
(more on this later). You<st1:PersonName w:st="on">'</st1:PersonName>d also see
that it references the basic ecj param file as well. But for now we<st1:PersonName
w:st="on">'</st1:PersonName>re only concerned with the parameters that are set
for this simple example.&nbsp;</span></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>Next we see that the parameter <b>eval.problem</b> is set
to&nbsp;<code><b><span style='font-size:10.0pt'>ec.app.gep.test1.test1</span></b></code>.
This is the name of the class that contains the code we need to supply for our
problem. We<st1:PersonName w:st="on">'</st1:PersonName>ll look at this in more
detail below.&nbsp;</p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>The next 2
parameters identify the maximum number of generations (in this case 100) that
we should run when trying to find a solution to the problem and whether we
should quit when we find a <st1:PersonName w:st="on">'</st1:PersonName>suitable<st1:PersonName
w:st="on">'</st1:PersonName> solution. In this case we say that we should quit
when the solution is found, rather than continuing until 100 generations have
been run. The parameter pop.subpop.0.size tells how many instances of possible
models (or chromosomes or individuals in gep terminology) we should try in each
generation. In this example we use 50. &nbsp;At this time a gep population
consists of only 1 subpopulation so we will always use pop.subpop.0 to refer to
the population.</span></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal style='margin-bottom:12.0pt'><span lang=EN style='mso-ansi-language:
EN'>The following <span class=GramE>set</span> of parameters that start with
gep.species identify various things that are very specific to gep problems.
Each individual in the population is a chromosome that can be composed of 1 or
more genes (there is an expectation that you<st1:PersonName w:st="on">'</st1:PersonName>ve
taken the time to read external documents that describe the gep approach and
that you will be familiar with the concepts at this point; this documentation
will refer to these concepts without great explanations). Each gene is
identified by its <st1:PersonName w:st="on">'</st1:PersonName>head<st1:PersonName
w:st="on">'</st1:PersonName> size and if there are multiple genes one needs to
specify the operator that will link the gene expressions together. Here we use
chromosomes with 2 genes, each with a head size of 5 and linked by addition.
(NOTE: unlike Ferreira’s implementation, any operator can be used as the
linking function and not just +, -, * and / or ‘or’, ‘and’, ‘xor’, ‘nor’, and
‘nand’ for logical problems. Most commonly the linking operator is addition. If
operators other than these standard ones are used then the type of operator
(logical or non-logical) must be correct and the arity must agree with the
number of genes in a chromosome).</span></p>

<p class=MsoNormal>There are 4 types of problems that can be handled by the gep
approach (at this time). In this case we are doing a function finding problem.
Genes can be composed of various symbols that represent the functions and
terminals (independent variables) that can be used. There can also be constants
but more on that later. We need to identify these symbols so that the system
can build appropriate genes. The functions are identified by telling how many
there are (functionsize) and what they are (function.0, function.1, etc.). In
this problem we have decided to allow the genes to use the 4 functions:
addition, subtraction, multiplication and addition. There are a very large
number of supported functions. These can be found in the ec/gep/symbols
directory. It is quite simple to add more functions as needed. Just follow the
example of a similar type of function and create a new class. The terminal or
independent variable symbols can also be defined in the parameter file but for
this example we have a training data file that contains the terminal symbols in
the 1st row of the file. So we specify that gep.species.symbolset.terminalfilename
is&nbsp;<code><b><span style='font-size:10.0pt'>ec/app/gep/test1/test1.txt</span></b></code>.
As shown above this will identify the 2 terminal symbols x and y (as well as
the dependent variable symbol z).&nbsp;</p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal style='margin-bottom:12.0pt'><span lang=EN style='mso-ansi-language:
EN'>The final entries in the parameter file describe what we want in the way of
output from the system. This is called the statistics output. In this example
we use the class ec.gep.GEPSimpleStatistics to generate our output and specify
that the output will go to the file stats1.txt. Users may want to build their
own statistics classes to control the type of information that is generated.</span></p>

<p class=MsoNormal>Finally we need to provide a simple program that tells the
system how to evaluate the <b>fitness</b> of chromosomes so decisions can be
made as the individuals are discarded or moved to the next generation (possibly
with modifications -- mutations, inversions, and various forms of recombination
and transposition). The code for the class ec.gep.test.test1 is shown below.</p>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
<code><span style='font-size:10.0pt'>import ec.*;</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>import ec.gep.*;</code><br>
<code>import ec.simple.*;</code><br>
<br>
<code>/**</code><br>
<code>&nbsp;*&nbsp; The problem is to find the equation x*x + y </code><br>
<code>&nbsp;*&nbsp; </code><br>
<code>&nbsp;*&nbsp; Data and variable names are in the data file test1.txt</code><br>
<code>&nbsp;*/</code><br>
<code>public class test1 extends GEPProblem implements SimpleProblemForm </code><br>
<code>{</code><br>
<code>&nbsp;&nbsp;static double IDEAL_FITNESS_MINIMUM = 999.9999;</code><br>
<code>&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp; public void evaluate(EvolutionState state, Individual ind, int
threadnum) </code><br>
<code>&nbsp;&nbsp;{</code><br>
<code>&nbsp;&nbsp;&nbsp; if (!ind.evaluated)&nbsp; // don<st1:PersonName w:st="on">'</st1:PersonName>t
bother reevaluating</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;{</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; // Mean Squared Error (MSE) fitness is
normalized between 0 and 1000 (1000 * (1/(1+MSE))</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; double fitness =
GEPFitnessFunction.MSEfitness(true, (GEPIndividual)ind);</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; // the fitness better be SimpleFitness!</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; SimpleFitness f = ((SimpleFitness<span
class=GramE>)ind.fitness</span>);</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; f.setFitness(state,(float)fitness, fitness
&gt;= IDEAL_FITNESS_MINIMUM);</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ind.evaluated = true;</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;if (fitness &gt;= IDEAL_FITNESS_MINIMUM)</code><br>
<code>&nbsp;&nbsp; &nbsp;&nbsp; {&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;((GEPIndividual)<st1:place
w:st="on"><st1:State w:st="on">ind</st1:State></st1:place>).printIndividualForHumans(state,
1, 1);</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp; }&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;}</code><br>
<code>&nbsp; }</code><br>
<code>}</code></span><span style='font-size:10.0pt'> </span></p>

<p class=MsoNormal>This is the basic format that should be used in the user
program. The outer test just says that if we<st1:PersonName w:st="on">'</st1:PersonName>ve
already evaluated the fitness of the individual (chromosome/model) then don<st1:PersonName
w:st="on">'</st1:PersonName>t do anything. Otherwise use one of the fitness
functions to evaluate the model (in this case we use Mean Squared Error), set
the fitness indicating whether it is an ideal solution to the problem and then
indicate that the model was evaluated. Printing the model is not required but
was done in this case. It will be captured in the statistics file output. There
are a number of possible fitness functions to choose from (see
ec.gep.GEPFitnessFunction class) and you can easily create your own.&nbsp;</p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>Note: the 1<sup>st</sup> argument to a fitness function
evaluation is a <span class=GramE>boolean</span> that specifies whether
training data (true) or testing data (false) should be used when calculating
the fitness of the function. The training data will always be used when
measuring the fitness of functions (models) created during the evolutionary
process. The fitness of the model with testing data will normally be used to
validate that ‘good’ models selected from the training (evolutionary) exercise
are still good when subjected to other data.<span style='mso-spacerun:yes'> 
</span></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal style='margin-bottom:12.0pt'>So now we have all of the
required pieces to define our problem and to execute it we can use the command
line or the ECJ console. To run from a command line execute something like<span
class=GramE>:</span><br>
<br>
<code><span style='font-size:10.0pt'>D:\ecj&gt;java -classpath
./;./start/javacsv.jar ec.Evolve -file ec/app/gep/test1/test1.params</span></code></p>

<p class=MsoNormal style='margin-bottom:12.0pt'>You<st1:PersonName w:st="on">'</st1:PersonName>ll
notice that there is a jar file called javacsv.jar that must be included in the
classpath. This is used to read the CSV files and can be found at
http://sourceforge.net/projects/javacsv/. Assuming all goes well you will see
the following output (or very similar output) on the console.</p>

<p class=MsoNormal><code><span style='font-size:10.0pt'>| ECJ</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>| An evolutionary computation system (version 15)</code><br>
<code>| By Sean Luke</code><br>
<code>| Contributors: L. Panait, G. Balan, S. Paus, Z. Skolicki,</code><br>
<code>|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
J. Bassett, R. Hubley, and A. Chircop</code><br>
<code>| URL: http://cs.gmu.edu/~eclab/projects/ecj/</code><br>
<code>| Mail: ecj-help@cs.gmu.edu</code><br>
<code>|&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; (better: join ECJ-INTEREST at URL
above)</code><br>
<code>| Date: April 4, 2006</code><br>
<code>| Current Java: 1.5.0_10 / Java HotSpot(TM) Client VM-1.5.0_10-b03</code><br>
<code>| Required Minimum Java: 1.3</code><br>
<br>
<code>Threads:&nbsp; breed/1 eval/1</code><br>
<code>Seed: 4357</code><br>
<code>Job: 0</code><br>
<code>Setting up</code><br>
<code>Initializing Generation 0</code><br>
<code>WARNING:</code><br>
<code>Weight for GEP Function must be &gt; 0; defaulting to 1)</code><br>
<code>PARAMETER: pop.subpop.0.species.symbolset.function.0.weight</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; ALSO: gep.species.symbolset.function.0.weight</code><br>
<code>WARNING:</code><br>
<code>Weight for GEP Function must be &gt; 0; defaulting to 1)</code><br>
<code>PARAMETER: pop.subpop.0.species.symbolset.function.1.weight</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; ALSO: gep.species.symbolset.function.1.weight</code><br>
<code>WARNING:</code><br>
<code>Weight for GEP Function must be &gt; 0; defaulting to 1)</code><br>
<code>PARAMETER: pop.subpop.0.species.symbolset.function.2.weight</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; ALSO: gep.species.symbolset.function.2.weight</code><br>
<code>WARNING:</code><br>
<code>Weight for GEP Function must be &gt; 0; defaulting to 1)</code><br>
<code>PARAMETER: pop.subpop.0.species.symbolset.function.3.weight</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; ALSO: gep.species.symbolset.function.3.weight</code><br>
<code>Generation 1</code></span></p>

<p class=MsoNormal><code><span style='font-size:10.0pt'>Evaluated: T</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>Fitness: 1000.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>*.x.x.y.y.x.x.y.x.x.x</code><br>
<code>Gene 1</code><br>
<code>y./.x.+.+.x.x.x.x.x.x</code><br>
<br>
<code>(x*x) + y</code><br>
<br>
<code>Found Ideal Individual</code></span><br>
<br>
<span class=GramE>The</span> warnings about weight for the functions can be ignored.
There is an option to identify weights for the functions so that the ones with
higher weights are selected more often when building and evolving genes. The
default of 1 was used by not specifying the weights. In the code we chose to
display&nbsp;the best individual when it was found and thus the output showing
the details for the chromosome with an ideal fitness of 1000. It identifies the
linking function used (+), the contents of the 2 genes (in Karva format ... see
the gep documentation in papers by Ferreira, etc.) and the expression of the
chromosome in a mathematical formula:<br>
<br>
&nbsp;&nbsp;&nbsp; (x*x) + y<br>
<br>
<span class=GramE>This</span> is clearly the solution we were looking for. You
will also find the statistics for the run in the file test1.stat. This should
contain the following information:</p>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
<code><span style='font-size:10.0pt'>Seed(s) used in this job: 4357</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>Problem Type: functionfinding</code><br>
<br>
<code>Generation: 0</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 274.6781</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>-.x.-.y.y.y.y.x.y.y.y</code><br>
<code>Gene 1</code><br>
<code>+.x.y.y.-.x.y.y.y.y.x</code><br>
<br>
<code>(x-(y-y)) + (x+y)</code><br>
<br>
<code>Generation: 1</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 1000.0&lt;small&gt;&lt;code&gt;</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>*.x.x.y.y.x.x.y.x.x.x</code><br>
<code>Gene 1</code><br>
<code>y./.x.+.+.x.x.x.x.x.x</code><br>
<br>
<code>(x*x) + y</code><br>
<br>
<code>Best Individual of Run:</code><br>
<code>Found at Generation: 1</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 1000.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>*.x.x.y.y.x.x.y.x.x.x</code><br>
<code>Gene 1</code><br>
<code>y./.x.+.+.x.x.x.x.x.x</code><br>
<br>
<code>(x*x) + y</code><br>
<br>
<code>Size of program: 4</code><br>
<code>Variables used(count): x(2), y(1)</code><br>
<br>
<code>***** TRAINING data results *****</code><br>
<code>Statistics: </code><br>
<code>MSE:&nbsp; 0.0&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; RAE:&nbsp;
0.0&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; RSE:&nbsp; 0.0</code><br>
<code>RMSE: 0.0&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; MAE:&nbsp;
0.0&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; RRSE: 0.0</code><br>
<br>
<code>Observed&nbsp;&nbsp;&nbsp; Computed</code><br>
<code>1.25&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1.25</code><br>
<code>3.0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 3.0</code><br>
<code>8.0&nbsp;&nbsp; &nbsp; &nbsp; &nbsp; 8.0</code><br>
<code>21.0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 21.0</code></span></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>The output shows
the best individual found in each generation. There were only 2 generations
needed to find a solution to this problem since it was a simple one. Notice
that in the first generation (generation 0), the best solution found was
(x-(y-y) + (x+y). It had a fitness value of 274.6781, well below the perfect
value of 1000 that we get with the best solution of generation 1. The
statistics also provide some other information. The size of the best solution
found during the run was 4. This means that there were 4 symbols used in the
expression (excluding the linking function). So we had 2 x terminal symbols, 1
y terminal symbol and the * function symbol in the solution. We also display some
statistical values including mean squared error (MSE), root mean squared error
(RMSE), etc. Finally we display the values expected or observed (from our
training data) and the values produced by the best model in the run, in this
case a perfect model.<o:p></o:p></span></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>Note that in more
recent versions of ECJ-GEP you can request that the mathematical expressions
found be simplified. In which case you would see in the results and statistics
file something like:<o:p></o:p></span></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal><span style='font-size:10.0pt;font-family:"Courier New"'>MATH<o:p></o:p></span></p>

<p class=MsoNormal><span style='font-size:10.0pt;font-family:"Courier New"'>(<span
class=GramE>y</span>+(x*x))<o:p></o:p></span></p>

<p class=MsoNormal><span style='font-size:10.0pt;font-family:"Courier New"'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal><span style='font-size:10.0pt;font-family:"Courier New"'>MATH
(SIMPLIFIED)<o:p></o:p></span></p>

<p class=MsoNormal><span class=GramE><span style='font-size:10.0pt;font-family:
"Courier New"'>x^</span></span><span style='font-size:10.0pt;font-family:"Courier New"'>2+y<o:p></o:p></span></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>But so note that in most cases it is best to do your
experiments without simplification turned on since the software system used to
do the simplification of expressions (meditor) has been know to cause crashes
or non terminating loops for some expressions. Also the meditor.jar file must
be in the ‘start’ directory. To turn on simplification the following line must
be in the parameter file:</p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><span class=GramE><span style='font-size:10.0pt;font-family:
"Courier New"'>pop.subpop.0.species.ind.simplify-expressions</span></span><span
style='font-size:10.0pt;font-family:"Courier New"'> = true<o:p></o:p></span></p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>So, that<st1:PersonName
w:st="on">'</st1:PersonName>s all there is to it. Let<st1:PersonName w:st="on">'</st1:PersonName>s
look at 2 more simple problems, test2 and test3, to see some minor variations
in the process. First the test2 example is solving a problem that should result
in a model with the expression x^4 + x^3 + x^2 + x. It has a single independent
variable, x. And in this example, instead of getting the terminal symbols and
data from a file we define the terminal symbols in the param file,
test2.param,&nbsp;and define the data in the program file, test2.java. The
relevant part of the param file is:</span></p>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
<code><span style='font-size:10.0pt'>gep.species.symbolset.terminalsize = 1</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>gep.species.symbolset.terminal.0 = x</code><br>
<code>gep.species.symbolset.functionsize = 4</code><br>
<code>gep.species.symbolset.function.0 = Add</code><br>
<code>gep.species.symbolset.function.0.weight = 1</code><br>
<code>gep.species.symbolset.function.1 = Sub</code><br>
<code>gep.species.symbolset.function.1.weight = 1</code><br>
<code>gep.species.symbolset.function.2 = Mul</code><br>
<code>gep.species.symbolset.function.2.weight = 1</code><br>
<code>gep.species.symbolset.function.3 = Div</code><br>
<code>gep.species.symbolset.function.3.weight = 1</code></span></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>Here we<st1:PersonName
w:st="on">'</st1:PersonName>ve identified that there is 1 terminal symbol, x.
We<st1:PersonName w:st="on">'</st1:PersonName>ve also explicitly specified that
the weight for each function should be 1. Specifying the data in the program is
a bit more complex but not much.</span></p>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
<code><span style='font-size:10.0pt'>package ec.gep.test;</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>import ec.*;</code><br>
<code>import ec.gep.*;</code><br>
<code>import ec.simple.*;</code><br>
<br>
<code>public class test2 extends GEPProblem implements SimpleProblemForm </code><br>
<code>{</code><br>
<code>&nbsp;&nbsp;&nbsp; public static final double IDEAL_FITNESS_MINIMUM =
0.9999999999;</code><br>
<br>
<code>&nbsp;&nbsp;&nbsp; public double xvalues[] = { 9.500366, -6.130432,
3.252685, 7.88797, 9.090484,</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; 1.485199,
-3.950531, 10.003326, -0.607453, 5.469299};&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp; public double zvalues[] = { 9103.54918799079,
1213.47815867463, 160.18146840485, </code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
4432.23529247904, 7671.79392962917, 11.8327154232663, 193.570365560718, </code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
11124.3786278048, -0.326443121119579, 1093.78835980998};&nbsp; // x^4 + x^3 +
x^2 + x</code><br>
<code>&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp; public double[] getDataValues( String label )</code><br>
<code>&nbsp;&nbsp;&nbsp; {</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; if (label.equals(&quot;x&quot;))</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; return
(xvalues);</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; else if
(label.equals(&quot;dependentVariable&quot;)) // always called <st1:PersonName
w:st="on">'</st1:PersonName>dependentVariable<st1:PersonName w:st="on">'</st1:PersonName></code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; return
(zvalues);</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; else</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; return
null;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp; }</code><br>
<br>
<code>&nbsp;&nbsp;&nbsp; public void evaluate(EvolutionState state, Individual
ind, int threadnum) </code><br>
<code>&nbsp;&nbsp;&nbsp; {</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; if (!ind.evaluated)&nbsp; //
don<st1:PersonName w:st="on">'</st1:PersonName>t bother reevaluating</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; {</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; //
Mean Squared Error (MSE) fitness is normalized between 0 and 1000 (1000 *
(1/(1+MSE))</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; double
fitness = GEPFitnessFunction.MSEfitness((GEPIndividual)ind);</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; // the
fitness better be SimpleFitness!</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
SimpleFitness f = (true, (SimpleFitness<span class=GramE>)ind.fitness</span>);</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; f.setFitness(state,(float)fitness,
fitness &gt;= IDEAL_FITNESS_MINIMUM);</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
ind.evaluated = true;</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; if (fitness &gt;=
IDEAL_FITNESS_MINIMUM)</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; {&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
((GEPIndividual)<st1:place w:st="on"><st1:State w:st="on">ind</st1:State></st1:place>).printIndividualForHumans(state,
1, 1);</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; }&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; }</code><br>
<code>&nbsp;&nbsp;&nbsp; }</code><br>
<code>}</code></span></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>The main
difference here is that we have to provide a method getDataValues to supply the
data values for the terminal symbols identified in the param file. In this case
that is the x values. The ECJ system will call this function during
initialization to get the values, passing the symbol &quot;x&quot; as an
argument. The user program must provide a <span class=GramE>double[</span>]
result. In the previous example the data was in a file and the first row had
the symbols for the independent variable and the dependent variable. In the
parameter file we don<st1:PersonName w:st="on">'</st1:PersonName>t specify the
name of the dependent variable so the system will ask for the dependent
variable values using the label &quot;dependentVariable&quot;. It is probably
most common to use a file to hold the training data. <span class=GramE>However,
if the training data is to be provided by performing a calculation (via a
simulation or some other means for example), then using the user program to
provide the data might make sense.</span>&nbsp;&nbsp;For test2 the statistics
file will look like:</span></p>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
<code><span style='font-size:10.0pt'>Seed(s) used in this job: 4357</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>Problem Type: functionfinding</code><br>
<br>
<code>Generation: 0</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 0.004819467</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>-.-.-.x.*.*.-.x.x.x.x.x.x.x.x.x.x</code><br>
<code>Gene 1</code><br>
<code>*.+.-.x./.x././.x.x.x.x.x.x.x.x.x</code><br>
<code>Gene 2</code><br>
<code>*.+.+.x.*.*.x.x.x.x.x.x.x.x.x.x.x</code><br>
<br>
<code>((x-(x*x))-((x*x)-(x-x))) + ((x+((x/x)/x))*(x-(x/x))) +
((x+(x*x))*((x*x)+x))</code><br>
<br>
<code>Generation: 1</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 410.7618</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>-.-.*.x.*.*.-.x.x.x.x.x.x.x.x.x.x</code><br>
<code>Gene 1</code><br>
<code>*.+.-.x./.x././.x.x.x.x.x.x.x.x.x</code><br>
<code>Gene 2</code><br>
<code>*.+.+.x.*.*./.x.x.x.x.x.x.x.x.x.x</code><br>
<br>
<code>((x-(x*x))-((x*x)*(x-x))) + ((x+((x/x)/x))*(x-(x/x))) +
((x+(x*x))*((x*x)+(x/x)))</code><br>
<br>
<code>Generation: 6</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 500.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>-.-.*.x.x.x.x.x.x.x.x.x.x.x.x.x.x</code><br>
<code>Gene 1</code><br>
<code>*.+./.x./.x././.x.x.x.x.x.x.x.x.x</code><br>
<code>Gene 2</code><br>
<code>*.+.+.x.*.*./.x.x.x.x.x.x.x.x.x.x</code><br>
<br>
<code>((x-x)-(x*x)) + ((x+((x/x)/x))*(x/(x/x))) + ((x+(x*x))*((x*x)+(x/x)))</code><br>
<br>
<code>Generation: 14</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 1000.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>-.-.*.x.x.x.x.x.x.x.x.x.x.x.x.x.x</code><br>
<code>Gene 1</code><br>
<code>*.x.x.x.x.x.x.+.x.x.x.x.x.x.x.x.x</code><br>
<code>Gene 2</code><br>
<code>*.+.+.x.*.*./.x.x.x.x.x.x.x.x.x.x</code><br>
<br>
<code>((x-x)-(x*x)) + (x*x) + ((x+(x*x))*((x*x)+(x/x)))</code><br>
<br>
<code>Best Individual of Run:</code><br>
<code>Found at Generation: 14</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 1000.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>-.-.*.x.x.x.x.x.x.x.x.x.x.x.x.x.x</code><br>
<code>Gene 1</code><br>
<code>*.x.x.x.x.x.x.+.x.x.x.x.x.x.x.x.x</code><br>
<code>Gene 2</code><br>
<code>*.+.+.x.*.*./.x.x.x.x.x.x.x.x.x.x</code><br>
<br>
<code>((x-x)-(x*x)) + (x*x) + ((x+(x*x))*((x*x)+(x/x)))</code><br>
<br>
<code>Size of program: 23</code><br>
<code>Variables used(count): x(13)</code><br>
<br>
<code>***** TRAINING data results *****</code><br>
<code>Statistics: </code><br>
<code>MSE:&nbsp; 7.182716450022835E-24&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
RAE:&nbsp; 4.832470396077297E-16&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
RSE:&nbsp; 4.344516272977063E-31</code><br>
<code>RMSE: 2.680059038533076E-12&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
MAE:&nbsp; 1.7715995337397316E-12&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; RRSE:
6.591294465411982E-16</code><br>
<br>
<code>Observed&nbsp;&nbsp;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Computed</code><br>
<code>9103.54918799079&nbsp;&nbsp;&nbsp; 9103.549187990786</code><br>
<code>1213.47815867463&nbsp;&nbsp;&nbsp; 1213.4781586746294</code><br>
<code>160.18146840485&nbsp;&nbsp;&nbsp; &nbsp;160.18146840484988</code><br>
<code>4432.23529247904&nbsp;&nbsp;&nbsp; 4432.235292479036</code><br>
<code>7671.79392962917&nbsp;&nbsp;&nbsp; 7671.793929629167</code><br>
<code>11.8327154232663&nbsp;&nbsp;&nbsp; 11.832715423266338</code><br>
<code>193.570365560718&nbsp;&nbsp;&nbsp; 193.57036556071805</code><br>
<code>11124.3786278048&nbsp;&nbsp;&nbsp; 11124.378627804794</code><br>
<code>-0.326443121119579&nbsp;&nbsp;-0.3264431211195794</code><br>
<code>1093.78835980998&nbsp;&nbsp;&nbsp; 1093.7883598099795</code></span></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>In this case it
took 14 generations to solve the problem with an ideal (fitness = 1000)
solution. The solution could use some simplification so that we see the first
term is -x*x, the second is x*x (which cancel each other) and the third is
(x+x*x<span class=GramE>)(</span>x+1) or x^4+x^3+x^2+x, the desired solution.
The errors in the statistics are not exactly zero but are quite small. This is
due to loss of precision as the complex (not simplified) expression is
evaluated. It also demonstrates why we use double and not float for all
calculations. One final thing to note is that we don<st1:PersonName w:st="on">'</st1:PersonName>t
print the best result after each generation. We only printed the results when
there was an improvement in the fitness of the best of a generation (at
generations 1, 6 and 14). This can be controlled by a parameter for the
statistics. You can print the results for every generation, only when the
fitness improves (as we did here by default) or only the final generation of
the run (see the gep.params files for details). <o:p></o:p></span></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>Again, if
simplification of math expressions was turned on you would see something like
the following in the output:<o:p></o:p></span></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal><span lang=EN style='font-size:10.0pt;font-family:"Courier New";
mso-ansi-language:EN'>MATH<o:p></o:p></span></p>

<p class=MsoNormal><code><span style='font-size:10.0pt'>((x-x)-(x*x)) + (x*x) +
((x+(x*x))*((x*x<span class=GramE>)+</span>(x/x)))<o:p></o:p></span></code></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal><span lang=EN style='font-size:10.0pt;font-family:"Courier New";
mso-ansi-language:EN'>MATH (SIMPLIFIED)<o:p></o:p></span></p>

<p class=MsoNormal><span lang=EN style='font-size:10.0pt;font-family:"Courier New";
mso-ansi-language:EN'>x+x^2+x^3+x^4<o:p></o:p></span></p>

<p class=MsoNormal><span lang=EN style='font-size:10.0pt;font-family:"Courier New";
mso-ansi-language:EN'><br>
</span><span lang=EN style='mso-ansi-language:EN'><br>
Test3 is just test1 again, but this time we do not have to supply a Java
program at all, we<st1:PersonName w:st="on">'</st1:PersonName>ll use the
default one. The test3.param file has the following changes:</span></p>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
<code><span style='font-size:10.0pt'># Problem</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>#===============================</code><br>
<code>eval.problem = ec.gep.GEPDefaultUserProg</code><br>
<code>eval.problem.fitness-function = RH</code><br>
<code>eval.problem.fitness-function-arg0 = 2.0</code></span></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>Instead of
specifying some user program like ec.gep.test.test3 we use
ec.gep.GEPDefaultUserProg. But because we do this we need to identify which
fitness function to use. There are more than 25 fitness functions provided and
most require a single argument, the individual being evaluated. Some require
one or 2 extra numeric arguments. In the case of test3 we<st1:PersonName w:st="on">'</st1:PersonName>ve
used the Relative Hits (RH) fitness function. It requires one numeric argument,
the precision. It is not necessarily the best choice for this problem but does
illustrate the use of the parameters to specify the fitness function and an
argument. The RH function counts the number of hits (matches) between the expected
results in the training set and the computed results. A comparison is
considered to be a hit if the percentage difference between the expected and
computed values is less than the precision specified, in this case 2%. So the
test is:</span></p>

<p class=MsoNormal style='margin-bottom:12.0pt'><span style='font-size:10.0pt'><br>
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;</span><code><span style='font-size:10.0pt'>&nbsp;100*(<span
class=GramE>abs(</span>expected-computed)/expected) &lt; precision</span></code></p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>The maximum
number of hits possible is the number of sets of training data. The
GEPDefaultUserProg uses the RH function to determine fitness. Check out the
GEPFitnessFunction class to see the selection of functions available and to add
your own if so inclined. So for test3 we only had to create a parameter file
and a testing data file. The results of the run should look something like:</span></p>

<p class=MsoNormal style='margin-bottom:12.0pt'><span style='font-size:10.0pt'><br>
</span><code><span style='font-size:10.0pt'>Seed(s) used in this job: 4357</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>Problem Type: functionfinding</code><br>
<br>
<code>Generation: 0</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 3.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>*.-.-.+.y.y.x.x.y.x.x</code><br>
<code>Gene 1</code><br>
<code>-.y.-.y.y.y.y.y.x.y.x</code><br>
<br>
<code>(((x+y)-y)*(y-x)) + (y-(y-y))</code><br>
<br>
<code>Generation: 1</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 4.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>*.x.x.y.y.x.x.y.x.x.x</code><br>
<code>Gene 1</code><br>
<code>y./.x.+.+.x.x.x.x.x.x</code><br>
<br>
<code>(x*x) + y</code><br>
<br>
<code>Best Individual of Run:</code><br>
<code>Found at Generation: 1</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 4.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>*.x.x.y.y.x.x.y.x.x.x</code><br>
<code>Gene 1</code><br>
<code>y./.x.+.+.x.x.x.x.x.x</code><br>
<br>
<code>(x*x) + y</code><br>
<br>
<code>Size of program: 4</code><br>
<code>Variables used(count): x(2), y(1)</code><br>
<br>
<code>***** TRAINING data results *****</code><br>
<code>Statistics: </code><br>
<code>MSE:&nbsp; 0.0&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; RAE:&nbsp;
0.0&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; RSE:&nbsp; 0.0</code><br>
<code>RMSE: 0.0&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; MAE:&nbsp;
0.0&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; RRSE: 0.0</code><br>
<br>
<code>Observed&nbsp;&nbsp;&nbsp; Computed</code><br>
<code>1.25&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1.25</code><br>
<code>3.0&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; 3.0</code><br>
<code>8.0&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp; 8.0</code><br>
<code>21.0&nbsp;&nbsp; &nbsp; &nbsp;&nbsp; 21.0</code></span> </p>

<p class=MsoNormal><span lang=EN style='mso-ansi-language:EN'>As you can see
the final result is the same, we get the x*x+y equation that the data supports.
The maximum fitness value in this case was 4, since there were only 4 training
sets of data. <br>
<br>
This concludes the introduction to using the ECJ GEP extension. The next
section identifies a number of other examples that have been implemented,
showing more complex examples from the four types of problems: function
finding, logical, classification and time series. We<st1:PersonName w:st="on">'</st1:PersonName>ll
explain some of the particular parameters and considerations for these types of
problems as we briefly describe the examples.</span></p>

<p class=MsoNormal style='margin-bottom:12.0pt'><o:p>&nbsp;</o:p></p>

<h1><a name="Some_examples"><span class=section1msonormal>Some Other Examples</span></a></h1>

<h1><a name="Function_Finding_Examples"></a>Function Finding Examples</h1>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
These examples are similar to the test1, test2 and test3 examples we explored
in the previous section. We<st1:PersonName w:st="on">'</st1:PersonName>ll just
describe them very briefly and point out any interesting new things that are
introduced.<br>
<br>
<b>1. Acot Example</b> - this is an attempt to discover the ArcCotangent
function using a set of 100 training cases. There is only 1 gene, 30
individuals in the population and the RRSE (Root Relative Squared Error) fitness
function is used. There is one independent variable, d0, specified in the
Acot.param file. There are 17 functions used (+, -, *, /, Sqrt, Exp,
Floor,&nbsp;Ceiling, Abs, Neg, Sin, <st1:place w:st="on">Cos</st1:place>, Tan,
Asin, Acos, Atan, and Ln). <span class=GramE>The add</span>, subtract, multiply
and divide functions have a weight of 3 and the others a weight of 1. These are
just variations of things we<st1:PersonName w:st="on">'</st1:PersonName>ve seen
before. The new thing is that in this example we allow the genes to encode not
only functions and terminals but also constants. This is described in Ferreira<st1:PersonName
w:st="on">'</st1:PersonName>s book and the GeneXproSoft software. But in order
to use constants you do need to supply some extra information.<br>
<br>
<code><span style='font-size:10.0pt'>gep.species.use-constants&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
= true</span></code><span style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>gep.species.numconstantspergene&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; = 2</code><br>
<code>gep.species.integer-constants&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
= false</code><br>
<code>gep.species.constants-lowerlimit&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; = -10</code><br>
<code>gep.species.constants-upperlimit&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; = 10</code><br>
<code>gep.species.rnc-mutation-prob&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
= 0.01</code><br>
<code>gep.species.dc-mutation-prob&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
= 0.044</code><br>
<code>gep.species.dc-inversion-prob&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
= 0.1</code><br>
<code>gep.species.dc-istransposition-prob&nbsp;&nbsp; = 0.1</code></span><br>
<br>
First we identify that we are going to use constants and that there will be 2
constants allocated per gene. We are using floating point constants since the
integer-constants parameter is set to false. The values of the constants are
allowed to range between -10.0 and +10.0. The other parameters relate to how
constants evolve from generation to generation. This again is covered in
Ferreira<st1:PersonName w:st="on">'</st1:PersonName>s documentation. But these
values are the ones she suggests as defaults that work well for most problems.<br>
<br>
After running the example we get the output in the Acot.stat file.<br>
<br>
<code><span style='font-size:10.0pt'>Seed(s) used in this job: 4357</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>Problem Type: functionfinding</code><br>
<br>
<code>Generation: 0</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 499.85913</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>ceil.tan.C1.tan.d0.C0.tan.acos.d0.d0.d0.C1.C1.C1.d0.C1.C1</code><br>
<code>C0: -8.147180372252937</code><br>
<code>C1: 5.652730175877368</code><br>
<br>
<code>ceiling(tan(5.652730175877368))</code><br>
<br>
<code>Generation: 10</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 503.19714</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>/.cos.C1.tan.tan.sin.d0.d0.d0.C1.C0.C1.C0.C0.d0.C0.d0</code><br>
<code>C0: -1.9238205968468058</code><br>
<code>C1: 5.576872726631521</code><br>
<br>
<code>(cos(tan(tan(sin(d0))))/5.576872726631521)</code><br>
<br>
<code>Generation: 14</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 614.31683</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>/.cos.d0.floor.C1.d0.floor.d0.d0.C1.C0.C1.C1.C0.d0.C1.d0</code><br>
<code>C0: -8.147180372252937</code><br>
<code>C1: 5.652730175877368</code><br>
<br>
<code>(cos(floor(5.652730175877368))/d0)</code><br>
<br>
<code>Generation: 21</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 873.75574</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>sin././.cos./.neg.*.floor.d0.C1.C0.C0.C1.C1.d0.d0.d0</code><br>
<code>C0: -8.147180372252937</code><br>
<code>C1: 5.652730175877368</code><br>
<br>
<code>sin((((floor(5.652730175877368)/d0)/(-5.652730175877368))/cos(((-8.147180372252937)*(-8.147180372252937)))))</code><br>
<br>
<code>Generation: 366</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 1000.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>atan././.d0.d0.d0./.C1.d0.d0.C1.C0.C0.d0.C0.C1.C1</code><br>
<code>C0: -8.147180372252937</code><br>
<code>C1: -0.24815065988408236</code><br>
<br>
<code>atan(((d0/d0)/d0))</code><br>
<br>
<code>Best Individual of Run:</code><br>
<code>Found at Generation: 366</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 1000.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>atan././.d0.d0.d0./.C1.d0.d0.C1.C0.C0.d0.C0.C1.C1</code><br>
<code>C0: -8.147180372252937</code><br>
<code>C1: -0.24815065988408236</code><br>
<br>
<code><b>atan(((d0/d0)/d0))</b></code><br>
<br>
<code>Size of program: 6</code><br>
<code>Variables used(count): d0(3)</code><br>
<br>
<code>***** TRAINING data results *****</code><br>
<code>Statistics: </code><br>
<code>MSE:&nbsp; 6.5223928682607005E-31&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;RAE:&nbsp; 1.4189629717074661E-15&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp; RSE:&nbsp; 5.2194614261805544E-30</code><br>
<code>RMSE: 8.076133275436148E-16&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
MAE:&nbsp; 3.738676035425215E-16&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; RRSE:
2.284614065040429E-15</code><br>
<br>
<code>Observed&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Computed</code><br>
<code>-0.204002518946359&nbsp;&nbsp;&nbsp; -0.2040025189463591</code><br>
<code>-0.339365885046348&nbsp;&nbsp;&nbsp; -0.3393658850463476</code><br>
<code>-0.387522092113786&nbsp;&nbsp;&nbsp; -0.3875220921137857</code><br>
<code>-0.23995309898349&nbsp; &nbsp;&nbsp; -0.23995309898349046</code><br>
<code>0.515703851271719&nbsp;&nbsp; &nbsp;&nbsp; 0.5157038512717189</code><br>
<code>0.310364838997948&nbsp;&nbsp; &nbsp;&nbsp; 0.31036483899794765</code><br>
<code>&nbsp;&nbsp;&nbsp; ...</code><br>
<code>0.146141817137733&nbsp;&nbsp; &nbsp;&nbsp; 0.14614181713773297</code><br>
<code>-0.124890152806863&nbsp;&nbsp;&nbsp; -0.1248901528068634</code><br>
<code>-0.182368282843355&nbsp;&nbsp;&nbsp; -0.18236828284335482</code><br>
<code>0.160813070832659&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 0.1608130708326594</code><br>
<code>0.139880646214395&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 0.13988064621439456</code><br>
</span><br>
It took 366 generations in this case (with the seed provided) and found the
expression atan(((d0/d0)/d0)) which is equivalent to atan(1/d0), the definition
of acot(d0). There were improvements in the best fitness value at generations
10, 14, 21, and 366. So there was a lot of exploration of models through
evolution to improve between generation 21 and 366.<br>
<br>
<b>2. Data mining example</b> - This is an interesting variation of the test2
problem described above. This high-dimensional toy problem illustrates the use
of a great deal of extraneous&nbsp;information being filtered during evolution.
In this example 15 out of 16 independent&nbsp;variables are meaningless. As in
test2 the target function is the quartic polynomial f(x) = x^4+x^3+x^2+x. The
data contained 161 sets with 15 independent variable values plus the dependent
variable values. As we can see from the (partial) output shown below, it takes
1596 generations (with the given seed as the starting point) to get to the best
solution.<br>
<br>
<code><span style='font-size:10.0pt'>Seed(s) used in this job: 4357</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>Problem Type: functionfinding</code><br>
<br>
<code>Generation: 0</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 459.41364</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>+./.-.*.*./.-.*.d0.d2.d15.d13.d4.d12.d10.d6.d6</code><br>
<code>Gene 1</code><br>
<code>+.d6.*.-.+.d5./.*.d3.d12.d3.d13.d0.d8.d14.d12.d7</code><br>
<code>Gene 2</code><br>
<code>-.*.-./.-.d8.*./.d0.d13.d8.d7.d12.d0.d11.d4.d5</code><br>
<br>
<code>((((d6*d6)*d0)/(d2*d15))+((d13/d4)-(d12-d10))) +
(d6+((d5-(d12/d3))*((d13*d0)+d3))) + ((((d0/d11)/d0)*(d13-d8))-(d8-(d7*d12)))</code><br>
<br>
<code>Generation: 3</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 461.31992</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>+.-.-.*.*./.-.*.d0.d2.d15.d13.d4.d12.d10.d6.d6</code><br>
<code>Gene 1</code><br>
<code>+.d6.*.-.+.d5./.*.d3.d12.d3.d13.d0.d8.d14.d12.d7</code><br>
<code>Gene 2</code><br>
<code>-.*.-./.-.d8.*./.d0.d13.d8.d7.d12.d0.d11.d4.d5</code><br>
<br>
<code>((((d6*d6)*d0)-(d2*d15))+((d13/d4)-(d12-d10))) + (d6+((d5-(d12/d3))*((d13*d0)+d3)))
+ ((((d0/d11)/d0)*(d13-d8))-(d8-(d7*d12)))</code><br>
<br>
<code>&nbsp;&nbsp;&nbsp; ...</code><br>
<br>
<code>Generation: 1347</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 998.6959</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>+.*.d6.*.d0.d0.d0.-.d5.d12.d7.d5.d8.d4.d10.d5.d4</code><br>
<code>Gene 1</code><br>
<code>-.*.d1.d0.+.*.d0.*.d0.d0.d0.d10.d11.d8.d10.d5.d10</code><br>
<code>Gene 2</code><br>
<code>+.-.d0.d1.+.d6.d1.d1.d3.d6.d0.d8.d5.d8.d0.d2.d14</code><br>
<br>
<code>(((d0*d0)*d0)+d6) + ((d0*(((d0*d0)*d0)+d0))-d1) + ((d1-(d6+d1))+d0)</code><br>
<br>
<code>Generation: 1596</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 1000.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>+.-.d0.d1.d6.-.d1.*.d0.d12.d2.d9.d5.d6.d9.d12.d12</code><br>
<code>Gene 1</code><br>
<code>-.*.d1.+.d0.*.d0.*.d0.d0.d0.d7.d5.d15.d9.d12.d1</code><br>
<code>Gene 2</code><br>
<code>+.*.d6.*.*./.d0.d0.d0.d0.d0.d10.d5.d8.d5.d8.d7</code><br>
<br>
<code>((d1-d6)+d0) + (((((d0*d0)*d0)+d0)*d0)-d1) + ((((d0/d0)*d0)*(d0*d0))+d6)</code><br>
<br>
<code>Best Individual of Run:</code><br>
<code>Found at Generation: 1596</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 1000.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>+.-.d0.d1.d6.-.d1.*.d0.d12.d2.d9.d5.d6.d9.d12.d12</code><br>
<code>Gene 1</code><br>
<code>-.*.d1.+.d0.*.d0.*.d0.d0.d0.d7.d5.d15.d9.d12.d1</code><br>
<code>Gene 2</code><br>
<code>+.*.d6.*.*./.d0.d0.d0.d0.d0.d10.d5.d8.d5.d8.d7</code><br>
<br>
<code>(<b>(d1-d6)+d0) + (((((d0*d0)*d0)+d0)*d0)-d1) +
((((d0/d0)*d0)*(d0*d0))+d6)</b></code><br>
<br>
<code>Size of program: 27</code><br>
<code>Variables used(count): d0(11), d1(2), d6(2)</code><br>
<br>
<code>***** TRAINING data results *****</code><br>
<code>Statistics: </code><br>
<code>MSE:&nbsp; 7.99571803020474E-23&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
RAE:&nbsp; 1.0804622056054273E-15&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; RSE:&nbsp;
4.6576363300204115E-30</code><br>
<code>RMSE: 8.941877895724556E-12&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; MAE:&nbsp;
3.3320664912135643E-12&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; RRSE:
2.1581557705643983E-15</code><br>
<br>
<code>Observed&nbsp;&nbsp;&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Computed</code><br>
<code>1418.74707069507&nbsp;&nbsp;&nbsp; 1418.7470706950726</code><br>
<code>75.3238688522697&nbsp;&nbsp;&nbsp; 75.3238688522697</code><br>
<code>11019.4252335609&nbsp;&nbsp;&nbsp; 11019.425233560933</code><br>
<code>1526.50067316805&nbsp;&nbsp;&nbsp; 1526.500673168051</code><br>
<code>11309.522753932&nbsp;&nbsp;&nbsp; &nbsp;11309.522753931999</code><br>
<code>6755.18648304066&nbsp;&nbsp;&nbsp; 6755.186483040658</code><br>
<code>&nbsp;&nbsp;&nbsp; ...</code><br>
<code>537.603328324592&nbsp;&nbsp;&nbsp; 537.603328324592</code><br>
<code>533.236668818159&nbsp;&nbsp;&nbsp; 533.2366688181589</code><br>
<code>3742.26191483226&nbsp;&nbsp;&nbsp; 3742.261914832265</code><br>
<code>6664.98407320731&nbsp;&nbsp;&nbsp; 6664.984073207315</code><br>
<code>8312.11655791031&nbsp;&nbsp;&nbsp; 8312.116557910313</code></span><br>
<br>
You<st1:PersonName w:st="on">'</st1:PersonName>ll also notice that for this
particular run the solution looks pretty complex but it simplifies to:<br>
<br>
&nbsp;&nbsp;&nbsp;<code><span style='font-size:10.0pt'> d1-d6+d0 + d0^4+d0^2-d1
+ d0^3+d6</span></code><br>
or<br>
&nbsp;&nbsp;&nbsp;<code><span style='font-size:10.0pt'> d0&nbsp;+ d0^4 + d0^2 +
d0^3</span></code><br>
<br>
the expected solution.&nbsp;There is a second version of the program,
dataMining2 that has the data in a CSV text file.<br>
<br>
<b>3. Production example</b> - In this example we have real-world production
data from 446 firms. The goal is to explain the Production&nbsp;as a function of
Labor, Material, and Capital. There is nothing new in the parameter file that
we haven<st1:PersonName w:st="on">'</st1:PersonName>t seen before. The data is
provided in the user program. The only unique thing we haven<st1:PersonName
w:st="on">'</st1:PersonName>t seen before is that the program runs to
completion at 3000 generations without reaching an <st1:PersonName w:st="on">'</st1:PersonName>ideal<st1:PersonName
w:st="on">'</st1:PersonName> or <st1:PersonName w:st="on">'</st1:PersonName>perfect<st1:PersonName
w:st="on">'</st1:PersonName> solution. This is probably more like real problems
with real (imperfect) data. We only explored the solution using the functions
add, subtract, multiply and divide. Perhaps there are more complicated
relations ships and trying some trigonometric functions might help. In any
case, this attempt provided a solution with a fitness of 8878.4472. In real
investigations we might run this problem 500 or 1000 times with different seeds
for 10000 generations each to try to find a set of best possible solutions. The
best solution of the run is extracted from the statistics output file and shown
below. Note the role of constants in the solution.<br>
<br>
<code><span style='font-size:10.0pt'>Best Individual of Run:</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>Found at Generation: 2999</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 878.4472</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>/.*./.-.-./.material./.material.capital.material.C0.labour.material.material.C0.C0</code><br>
<code>C0: 7.249937058678704</code><br>
<code>C1: -0.5794351361736538</code><br>
<br>
<code>Gene 1</code><br>
<code>material.material.material./.C0.+.labour.labour.capital.capital.labour.C1.material.material.capital.capital.C1</code><br>
<code>C0: 3.697751463839939</code><br>
<code>C1: 2.8929911331385565</code><br>
<br>
<code>Gene 2</code><br>
<code>*./././.-.C1./.-.material.C0.material.C1.material.material.labour.labour.labour</code><br>
<code>C0: -3.2099741094356737</code><br>
<code>C1: -0.49638400660778004</code><br>
<br>
<code>((((material/material)-material)*(capital-material))/((7.249937058678704/labour)/material))
+ material +
((((material-labour)/material)/((-3.2099741094356737)-material))*((-0.49638400660778004)/((-0.49638400660778004)/material)))</code><br>
<br>
<code>Size of program: 31</code><br>
<code>Variables used(count): labour(2), material(10), capital(1)</code></span></p>

<p class=MsoNormal style='margin-bottom:12.0pt'><b><span lang=EN
style='mso-ansi-language:EN'>4. &nbsp;Analog Circuit Design Example</span></b><span
lang=EN style='mso-ansi-language:EN'> - The goal is to find the transfer
function expressing the yield of an analog circuit&nbsp;in terms of three
parameter tolerances. The training set, consisting of 40 pairs&nbsp;of
tolerances and their corresponding yields, was obtained from n runs of <st1:place
w:st="on">Monte Carlo</st1:place>&nbsp;simulations. The source of the problem
is: <o:p></o:p></span></p>

<div style='margin-left:30.0pt'>

<p class=MsoNormal><span class=GramE><span lang=EN style='mso-ansi-language:
EN'>Zielinski, L. and J. Rutkowski, 2004.</span></span><span lang=EN
style='mso-ansi-language:EN'> Design Tolerancing with Utilization of Gene<br>
Expression Programming and Genetic Algorithm. <span class=GramE>In Proceedings
of the International <br>
Conference on Signals and Electronic Systems, <st1:place w:st="on"><st1:City
 w:st="on">Poznan</st1:City>, <st1:country-region w:st="on">Poland</st1:country-region></st1:place>.</span><o:p></o:p></span></p>

</div>

<p class=MsoNormal><br>
Again this is similar to other function finding problems we<st1:PersonName
w:st="on">'</st1:PersonName>ve seen. However, this problem sets parameters to:
use 10 integer constants per gene; run to 5000 generations;&nbsp;weight add,
subtract and multiply at 3; and to use ln, sqrt, and <span class=GramE>pow</span>
functions. Check out the files analogCircuit.java, analogCircuit.param and
analogCircuit.stat.<br>
<code><span style='font-size:10.0pt'>&nbsp;&nbsp;&nbsp; </span></code></p>

<h1><a name="Classification_Examples"></a>Classification Examples</h1>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
Classification examples we haven<st1:PersonName w:st="on">'</st1:PersonName>t
seen yet. Typically we<st1:PersonName w:st="on">'</st1:PersonName>ll have a set
of examples of things that belong to various classes. Each of the examples will
have a set of attributes that we hope allow us to predict which class the
examples belong to and as new (unclassified) examples <span class=GramE>arrive</span>
we can use this knowledge to determine the class of the new example. In the gep
approach we work with one class at a time by taking the training examples we
have and identifying the selected class if the example belongs to the class or
does not belong to the class. In this method, the dependent variable will have
the values 0 and 1, representing whether the example does not or does belong to
the class. By repeating for all of the classes we can construct a set of
equations that allows up to predict whether a new example belongs to each class
or not. The model (equation) produced will need to result in a zero or a 1. To
do this we select a threshold value. If the equations results in a value
greater than this value then return a 1 (in the class) and if below the
threshold return a 0 (not in the class). Let<st1:PersonName w:st="on">'</st1:PersonName>s
look at a few examples and see how to set things up in the parameter
files.&nbsp;<br>
<br>
<b>1. Iris Virginica Example</b> - In this problem the goal is to classify
three different types of irises based&nbsp;on four measurements: sepal length,
sepal width, petal length, and petal width.&nbsp;The original iris dataset
contains fifty examples each of three types <span class=GramE>of &nbsp;iris</span>:
Iris setosa, Iris versicolor, and Iris virginica. Here the sub-problem
Virginica&nbsp;versus Not Virginica is analyzed, where 100 randomly chosen
samples are used for training&nbsp;and the remaining 50 for testing. The new
things in the parameter file (IrisVirginica.param) are shown below.<br>
<br>
<code><span style='font-size:10.0pt'># Problem type must be one of:
functionfinding, classification, timeseries, logical</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code># Set default to be <st1:PersonName w:st="on">'</st1:PersonName>unknown<st1:PersonName
w:st="on">'</st1:PersonName> so user is forced to specify in the problem params
file </code><br>
<code>gep.species.problemtype = classification</code><br>
<br>
<code># if the problem is a classification type problem a threshold value (used</code><br>
<code># to convert real values to 0 or 1 during fitness calculations) must be </code><br>
<code># specified (do not specify this unless it IS a classification type
problem).</code><br>
<code>gep.species.classification-threshold<span class=GramE>&nbsp; =</span> 0.5</code><br>
<br>
<code>gep.species.symbolset.terminalfilename =
ec/app/gep/IrisVirginica/IrisVirginica.txt</code><br>
<code>gep.species.symbolset.testingdatafilename =
ec/app/gep/IrisVirginica/IrisVirginicaTest.txt</code></span></p>

<p class=MsoNormal>We see that the problem type is set to <b>classification</b>.
With classification problems we expect to see a classification threshold
specified, In this case the value is set to 0.5. The data is in CSV files. The
training data set is in IrisVirginica.txt. There are 100 sets of data in the
file along with the first record that holds the names of the terminals
(variables) as d0, d1, d2, and D.V. We see also that there is a testing data
set file specified as IrisVirginicaTest.txt. This file holds the 50 testing sets
of data and <b>does not</b> have the names of the variables in the first row!
When the best model is determined from the run, the gep system will
automatically access the test data and compare the results of using the model
with this data against the expected values. This will help in the assessment of
the quality of the model that was chosen. Let<st1:PersonName w:st="on">'</st1:PersonName>s
look at the output from a run (IrisVirginica.stat). We set the problem to run
for 5000 generations.</p>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
<code><span style='font-size:10.0pt'>Seed(s) used in this job: -1368642168</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>Problem Type: classification</code><br>
<code>Classification rounding threshold: 0.5</code><br>
<br>
<code>Generation: 0</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 801.24774</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>*./.+.+.*./.-.d2.d3.d0.d1.d3.d3.d3.d1.d3.d3</code><br>
<code>C0: 4.19396659531718</code><br>
<code>C1: 4.2865486803954775</code><br>
<br>
<code>Gene 1</code><br>
<code>+./.d3.-.+./.d2.d1.d1.d3.d0.C1.d2.d1.C1.d2.d1</code><br>
<code>C0: 2.4700381059064878</code><br>
<code>C1: 8.534256473201435</code><br>
<br>
<code>Gene 2</code><br>
<code>/.d1.*.-.*.d1.d2.d1.d0.d2.d2.d0.C1.d1.d0.C0.C0</code><br>
<code>C0: 4.031580228177383</code><br>
<code>C1: 1.4026131599950642</code><br>
<br>
<code>(((d2+d3)/(d0*d1))*((d3/d3)+(d3-d1))) + ((((d3/d0)-d2)/(d1+d1))+d3) +
(d1/((d1-d2)*(d1*d0)))</code><br>
<br>
<code>&nbsp;&nbsp;&nbsp; ...</code><br>
<br>
<code>Generation: 9</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 969.69696</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>-.d3./.+.d2.*./.d0.d1.d1.d1.d1.d3.d1.d2.d2.d2</code><br>
<code>C0: 0.2103929964396789</code><br>
<code>C1: 4.2865486803954775</code><br>
<br>
<code>Gene 1</code><br>
<code>-.*./.-.d3.d3.C1.d2.d1.d3.d2.C1.d0.d2.d3.d2.d1</code><br>
<code>C0: 2.4700381059064878</code><br>
<code>C1: 8.534256473201435</code><br>
<br>
<code>Gene 2</code><br>
<code>/.d1.*.-.*.d1.d2.d1.C1.d1.d3.d3.C0.d3.d0.C1.C0</code><br>
<code>C0: 4.031580228177383</code><br>
<code>C1: 1.4026131599950642</code><br>
<br>
<code>(d3-(((d0*d1)+(d1/d1))/d2)) + (((d2-d1)*d3)-(d3/8.534256473201435)) +
(d1/((d1-d2)*(d1*1.4026131599950642)))</code><br>
<br>
<code>Generation: 421</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 984.8485</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>-.-././.d3.d0.d3.*.d1.d1.d2.C0.d2.d3.C0.d2.d2</code><br>
<code>C0: 9.274259134965515</code><br>
<code>C1: 4.873682980602775</code><br>
<br>
<code>Gene 1</code><br>
<code>-.d3./.+.d2.*.C1.d0.d1.d2.d2.d2.d2.C0.d2.C1.d0</code><br>
<code>C0: 3.9855714658085004</code><br>
<code>C1: 8.534256473201435</code><br>
<br>
<code>Gene 2</code><br>
<code>-.d2././.d2.d2.C1.-.C0.C0.d1.d0.d0.d0.C1.d3.d0</code><br>
<code>C0: 1.8621216876649604</code><br>
<code>C1: 6.2762163390852646</code><br>
<br>
<code>((((d1*d2)/d1)-d3)-(d0/d3)) + (d3-(((d0*d1)+8.534256473201435)/d2)) +
(d2-((d2/6.2762163390852646)/d2))</code><br>
<br>
<code>Best Individual of Run:</code><br>
<code>Found at Generation: 421</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 984.8485</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>-.-././.d3.d0.d3.*.d1.d1.d2.C0.d2.d3.C0.d2.d2</code><br>
<code>C0: 9.274259134965515</code><br>
<code>C1: 4.873682980602775</code><br>
<br>
<code>Gene 1</code><br>
<code>-.d3./.+.d2.*.C1.d0.d1.d2.d2.d2.d2.C0.d2.C1.d0</code><br>
<code>C0: 3.9855714658085004</code><br>
<code>C1: 8.534256473201435</code><br>
<br>
<code>Gene 2</code><br>
<code>-.d2././.d2.d2.C1.-.C0.C0.d1.d0.d0.d0.C1.d3.d0</code><br>
<code>C0: 1.8621216876649604</code><br>
<code>C1: 6.2762163390852646</code><br>
<br>
<code>((((d1*d2)/d1)-d3)-(d0/d3)) + (d3-(((d0*d1)+8.534256473201435)/d2)) +
(d2-((d2/6.2762163390852646)/d2))</code><br>
<br>
<code>Size of program: 27</code><br>
<code>Variables used(count): d0(2), d1(3), d2(5), d3(3)</code><br>
<br>
<code>***** TRAINING data results *****</code><br>
<code>Confusion Matrix: </code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Predicted Value</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|&nbsp; Yes&nbsp;&nbsp; |&nbsp; No</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|----------------</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Yes|&nbsp; 34&nbsp;&nbsp;&nbsp; |&nbsp; 0</code><br>
<code>Actual Value&nbsp;&nbsp; |----------------</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; No
|&nbsp; 1&nbsp;&nbsp; &nbsp;&nbsp;|&nbsp; 65</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|----------------</code><br>
<br>
<br>
<code>***** TEST data results *****</code><br>
<code>Number of Calculation Errors: 0 out of 50 test sets</code><br>
<code>Confusion Matrix: </code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Predicted Value</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|&nbsp; Yes&nbsp;&nbsp; |&nbsp; No</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|----------------</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Yes|&nbsp; 16&nbsp;&nbsp;&nbsp; |&nbsp; 0</code><br>
<code>Actual Value&nbsp;&nbsp; |----------------</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; No
|&nbsp; 0&nbsp;&nbsp; &nbsp; |&nbsp; 34</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|----------------</code><br>
<br>
<code>Observed&nbsp;&nbsp;&nbsp; Computed</code><br>
<code>1.0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1.0</code><br>
<code>1.0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1.0</code><br>
<code>0.0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 0.0</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ...</code></span></p>

<p class=MsoNormal>We see from the <st1:PersonName w:st="on">'</st1:PersonName>confusion
matrix<st1:PersonName w:st="on">'</st1:PersonName> that the model produced from
this run predicted all but 1 data set correctly for the training data (a false
positive). And for the test data set it&nbsp;predicted all of the examples
correctly.&nbsp; We can also see that the output has been truncated. In the
stat file you<st1:PersonName w:st="on">'</st1:PersonName>ll see that all of the
observed and computed values are displayed, including the results for
the&nbsp;test data values (after the training data values). One last note: the
line that identifies the number of calculation errors refers to the number of
times that the model produced an arithmetic error when applied to the test set
of data. This can happen, for example, when the model has a division and the
test values lead to a zero divisor.</p>

<p class=MsoNormal style='margin-bottom:12.0pt'><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><b>2. Breast Cancer Example</b> - In this example of
diagnosis of breast cancer the task is to classify a tumor as either benign (0)
or malignant (1) based on nine different&nbsp;cell analysis (clump thickness,
uniformity of cell size, uniformity of cell shape, marginal&nbsp;adhesion,
single epithelial cell size, bare nuclei, bland chromatin, normal nucleoli, and
mitoses). Data obtained from PROBEN1 (Prechelt, L., 1994. PROBEN1 - A set of
neural network&nbsp;benchmark problems and benchmarking rules. <span
class=GramE>Technical Report 21/94, Univ. <st1:place w:st="on"><st1:City w:st="on">Karlsruhe</st1:City>,
 <st1:country-region w:st="on">Germany</st1:country-region></st1:place>).</span>&nbsp;Both
the technical report and the data set cancer1 used here are available for
anonymous FTP&nbsp;from Neural Bench archive at <st1:place w:st="on"><st1:PlaceName
 w:st="on">Carnegie</st1:PlaceName> <st1:PlaceName w:st="on">Mellon</st1:PlaceName>
 <st1:PlaceType w:st="on">University</st1:PlaceType></st1:place> (machine
ftp.cs.cmu.edu, directory&nbsp;/afs/cs/project/connect/bench/contrib/prechelt)
and from machine ftp.ira.uka.de in directory&nbsp;/pub/neuron. The file name in
both cases is proben1.tar.gz.</p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>Nothing new here except that there is a large amount of data
(training and testing). Run the problem or check out the results in
BreastCancer.stat.</p>

<p class=MsoNormal style='margin-bottom:12.0pt'><o:p>&nbsp;</o:p></p>

<p class=MsoNormal><b>3. DNA MicroArrays Example</b> - The objective of this
problem is to distinguish between 2 types of leukemia, ALL (0)&nbsp;and AML
(1). There are 38 training data sets and 34 testing data sets. The major
difference of this example is the number of attributes. There are 7129
independent variables! One minor change from the other examples is that we
specified a tab as the separator in the data files rather than the default
comma (gep.species.symbolset.terminalfileseparator = tab). The data sets are
available at:
&nbsp;http://sdmc.lit.org.sg/GEDatasets/Data/ALL-AML_Leukemia.zip. Note that
these files have very long lines and many editors will have difficulty reading
or editing them. With the default seed value an ideal solution is found after
2924 generations. But as you can see from the output the model did not perform
well with the test data. A good model for the training data doesn<st1:PersonName
w:st="on">'</st1:PersonName>t mean we have a solution. More runs will be
required to find solutions that can match the test and training data well.</p>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
<code><span style='font-size:10.0pt'>Best Individual of Run:</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>Found at Generation: 2924</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 1000.0</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>-.d2452.*.d4739.*.*.+.*.d4572.d4791.d1108.d4210.d670.d4458.d1985.d5621.d5374</code><br>
<code>Gene 1</code><br>
<code>+.d6935.*.+.+.*.+.d4555.d221.d1529.d3911.d2841.d1838.d2397.d4492.d6959.d1045</code><br>
<code>Gene 2</code><br>
<code>*.-.*.*.d6801.-.*.*.d11.d749.d6727.d2841.d7054.d5751.d6694.d436.d6498</code><br>
<br>
<code>(d2452-(d4739*(((d4210*d670)*d4572)*(d4791+d1108)))) +
(d6935+(((d1529*d3911)+(d2841+d1838))*(d4555+d221))) +
((((d5751*d6694)*d11)-d6801)*((d749-d6727)*(d2841*d7054)))</code><br>
<br>
<code>Size of program: 41</code><br>
<code>Variables used(count): d11(1), d221(1), d670(1), d749(1), d1108(1),
d1529(1), d1838(1), d2452(1), d2841(2), d3911(1), d4210(1), d4555(1), d4572(1),
d4739(1), d4791(1), d5751(1), d6694(1), d6727(1), d6801(1), d6935(1), d7054(1)</code><br>
<br>
<code>***** TRAINING data results *****</code><br>
<code>Confusion Matrix: </code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Predicted Value</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|&nbsp; Yes&nbsp;&nbsp;&nbsp;|&nbsp; No</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|----------------</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Yes|&nbsp; 11&nbsp;&nbsp;&nbsp; |&nbsp; 0</code><br>
<code>Actual Value&nbsp;&nbsp; |----------------</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; No
|&nbsp; 0&nbsp;&nbsp; &nbsp; |&nbsp; 27</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|----------------</code><br>
<br>
<br>
<code>***** TEST data results *****</code><br>
<code>Number of Calculation Errors: 0 out of 34 test sets</code><br>
<code>Confusion Matrix: </code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Predicted Value</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|&nbsp; Yes&nbsp;&nbsp;&nbsp;|&nbsp; No</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|----------------</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Yes|&nbsp; 4&nbsp;&nbsp; &nbsp;&nbsp;|&nbsp; 10</code><br>
<code>Actual Value&nbsp;&nbsp; |----------------</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; No
|&nbsp; 2&nbsp;&nbsp; &nbsp;&nbsp;|&nbsp; 18</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|----------------</code></span></p>

<p class=MsoNormal>Also see the publication:<br>
<br>
&nbsp;&nbsp;&nbsp; &quot;Molecular Classification of Cancer: Class Discovery
and Class Prediction by <br>
&nbsp;&nbsp;&nbsp; Gene&nbsp;Expression Monitoring&quot;. Science, 286:531-537,
October 1999. </p>

<p class=MsoNormal style='margin-bottom:12.0pt'><o:p>&nbsp;</o:p></p>

<h1><a name="Time_Series_Examples"></a>Time Series Examples</h1>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal style='margin-bottom:12.0pt'>The special thing about time
series problems (as found in gep) is that the data is a single sequence of
data. From a training sequence of time series data we<st1:PersonName w:st="on">'</st1:PersonName>d
like to create a model that can predict new values (i.e. given a test sequence
of time series data what would the next value in the sequence be). To this end
the sequential time series data is preprocessed to create a set of training
data sets and a set of testing data sets. To do this the user must provide a
bit of information about how the processing should be done. The user must
specify 3 things in the param file: <br>
<br>
<b>timeseries-delay</b> - This integer value determines how to use the values
in the timeseries data. If the value is 1, then use every time series value, if
it is 2, then use every other one, etc.<br>
<br>
<span class=GramE><b>timeseries-embeddingdimension</b></span>&nbsp;- This
integr value determines the number of timeseries points to use&nbsp;as
independent variables when transforming the set of time series data.
Another&nbsp;data point is used as the dependent variable value. So the time
series <st1:PersonName w:st="on">'</st1:PersonName>raw<st1:PersonName w:st="on">'</st1:PersonName>
data consisting of a list of single values is processed by splitting the data into&nbsp;groups
(rows) of size embeddingdimension+1. From the end of the time series
data&nbsp;embeddingdimension+1 <span class=GramE>values</span> are chosen (if
delay is 1 all values are chosen, if it is 2 every other one is chosen). The
last value is the independent variable value.&nbsp;Then the next row is
selected by moving <st1:PersonName w:st="on">'</st1:PersonName>delay<st1:PersonName
w:st="on">'</st1:PersonName>&nbsp;values from the end and choosing another
embeddingdimension+1 <span class=GramE>values</span>. This is
repeated&nbsp;until no more sets of size embeddingdimension+1 can be
chosen.&nbsp;If this produces&nbsp;n sets of data then <b>timeseries-testingpredictions</b>
of them are used for testing and (n -&nbsp;<b>timeseries-testingpredictions</b>)
are used for training.<br>
<br>
<span class=GramE><b>timeseries-testingpredictions</b></span> - an integer that
specifies the number of sets of data to devote to testing.<br>
<br>
So if we had the time series data:<br>
<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code><span style='font-size:10.0pt'>
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21</span></code><br>
<br>
and delay was 1 and&nbsp;<b>timeseries-embeddingdimension</b> was 4 then we<st1:PersonName
w:st="on">'</st1:PersonName>d process the set into&nbsp;the following 17 data
sets. If&nbsp;<b>timeseries-testingpredictions</b> was 6 then the 1st
11&nbsp;would be used for training and the LAST 6 for testing.&nbsp;<br>
<br>
<code><span style='font-size:10.0pt'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
iv1 iv2 iv3 iv4&nbsp; dv</span></code><span style='font-size:10.0pt;font-family:
"Courier New"'><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1&nbsp;&nbsp;
2&nbsp;&nbsp; 3&nbsp;&nbsp; 4&nbsp;&nbsp; 5</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2&nbsp;&nbsp;
3&nbsp;&nbsp; 4&nbsp;&nbsp; 5&nbsp;&nbsp; 6</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 3&nbsp;&nbsp;
4&nbsp;&nbsp; 5&nbsp;&nbsp; 6&nbsp;&nbsp; 7</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
. . .</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 14&nbsp; 15&nbsp;
16&nbsp; 17&nbsp; 18</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 15&nbsp; 16&nbsp;
17&nbsp; 18&nbsp; 19</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 16&nbsp; 17&nbsp;
18&nbsp; 19&nbsp; 20</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 17&nbsp; 18&nbsp;
19&nbsp; 20&nbsp; 21</code></span></p>

<p class=MsoNormal style='margin-bottom:12.0pt'>If delay was 2 then 7 sets
would be formed as:</p>

<p class=MsoNormal style='margin-bottom:12.0pt'><code><span style='font-size:
10.0pt'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; iv1 iv2 iv3 iv4&nbsp; dv</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1&nbsp;&nbsp;
3&nbsp;&nbsp; 5&nbsp;&nbsp; 7&nbsp;&nbsp; 9</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 3&nbsp;&nbsp;
5&nbsp;&nbsp; 7&nbsp;&nbsp; 9&nbsp; 11</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
. . .</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 9&nbsp; 11&nbsp;
13&nbsp; 15&nbsp; 17</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 11&nbsp; 13&nbsp;
15&nbsp; 17&nbsp; 19</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 13&nbsp; 15&nbsp;
17&nbsp; 19&nbsp; 21</code></span><br>
<br>
&nbsp;Of course the user can provide the data already preprocessed into the 2
data sets as we have done before, but that could require a lot more effort. The
timeseries data file must not have any information regarding the names of the
independent variables, since these are automatically created.<br>
<br>
<b>1. Dow Jones Example</b> - This example looks at the monthly closings of the
Dow-Jones industrial index, Aug. 1968 - Aug. 1981. (Source:&nbsp;Hipel and
Mcleod (1994). Time Series Modelling of Water Resources and
Environmental&nbsp;Systems, Elsevier. The original time series data file DJ.DAT
came from:<br>
<br>
&nbsp;&nbsp;&nbsp; Hyndman, R.J. (n.d.) Time Series Data Library,
http://www.robhyndman.info/TSDL/. <br>
&nbsp;&nbsp;&nbsp; <span class=GramE>Accessed on June 3, 2006.</span><br>
<br>
For the integral time series used in this example, each value from the original
raw timeseries was replaced by its moving average using a smoothing period p =
10.<br>
<br>
Of interest in the example is that there are 40 functions used to build the
models and that the 3 timeseries values are set to:<br>
<br>
gep.species.timeseries-delay&nbsp;&nbsp; =&nbsp; 1<br>
gep.species.timeseries-embeddingdimension = 10<br>
gep.species.timeseries-testingpredictions = 10 <br>
<br>
Some of the statistics output:<br>
<br>
<code><span style='font-size:10.0pt'>Seed(s) used in this job: 4357</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>Problem Type: timeseries</code><br>
<br>
<code>Generation: 0</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 862.8322</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>gau.comp.*.v6.C1.*.v7.mul3.v6.C0.v3.v3.v6.v8.v8.v3.v5.v6.v6.v9.v6.v1.C0.v0.v4.v2.v0.v2.v9.v3.C1.v8.v9</code><br>
<code>C0: -1.8928414198253574</code><br>
<code>C1: -9.528899601444923</code><br>
<br>
<code>Gene 1</code><br>
<code>max2.X3.v9.C0.exp.-.v0.v9.v0.v8.v3.v9.v6.v2.v1.v7.v6.v6.v0.v2.v3.v7.v6.v7.v3.v3.v1.v8.v7.v3.v8.v3.v7</code><br>
<code>C0: 5.353542014548175</code><br>
<code>C1: -9.236307382044629</code><br>
<br>
<code>Gene 2</code><br>
<code>pi.v4.X5.v0.*.comp.min2.v7.v3.C1.v9.v3.v5.v1.v2.v9.v7.C0.v7.v8.v6.v5.v3.v8.v9.v1.v8.v0.v9.v2.v5.v0.v1</code><br>
<code>C0: 9.111412390259325</code><br>
<code>C1: 9.94532215973475</code><br>
<br>
<code>exp(-pow((1-(v6*(-9.528899601444923))),2)) +
max((5.353542014548175^3),v9) + pi</code><br>
<br>
<code>Generation: 2</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 866.1386</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>gau.comp.*.v7.C1.*.v7.mul3.v6.C0.v4.v3.v6.v8.v8.v3.v5.v6.v6.v9.v6.v5.C0.v0.v4.v2.v0.v2.v9.v3.C1.v8.C1</code><br>
<code>C0: -1.8928414198253574</code><br>
<code>C1: -9.528899601444923</code><br>
<br>
<code>Gene 1</code><br>
<code>max2.X3.v9.C0.exp.-.v0.v9.v0.v8.v3.v9.v6.v2.v1.v7.v6.v6.v0.v2.v3.v7.v6.v7.v3.v3.v1.v8.v7.v3.v8.v3.v7</code><br>
<code>C0: 5.353542014548175</code><br>
<code>C1: -9.236307382044629</code><br>
<br>
<code>Gene 2</code><br>
<code>gau.comp.*.v6.C1.*.v7.mul3.v6.C0.v1.v2.v6.v8.v3.v3.v5.v2.v6.v9.v6.v1.C1.v0.v4.v2.v0.v2.v9.v3.C0.v8.v9</code><br>
<code>C0: 9.111412390259325</code><br>
<code>C1: 9.94532215973475</code><br>
<br>
<code>exp(-pow((1-(v7*(-9.528899601444923))),2)) +
max((5.353542014548175^3),v9) + exp(-pow((1-(v6*9.94532215973475)),2))</code><br>
<br>
<code>&nbsp;&nbsp;&nbsp; ...</code><br>
<br>
<code>Generation: 238</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 934.1794</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>-.v9.v8.sech.v1.v7.v9.v4.v7.C1.v2.v0.v3.v8.v3.v1.v5.v8.v8.v2.v8.v6.v2.v9.v2.v2.v1.v7.v8.v7.v0.v0.C0</code><br>
<code>C0: -1.8928414198253574</code><br>
<code>C1: 4.140535916050434</code><br>
<br>
<code>Gene 1</code><br>
<code>sin.min3.avg2.v2.v9.v8.v5.v0.v5.v8.v2.v9.v8.v8.v3.v0.v5.v5.v3.C1.v2.v0.v6.v9.v9.v2.v7.v5.C1.v8.v7.v0.v6</code><br>
<code>C0: 7.570619662652462</code><br>
<code>C1: -1.542568610489699</code><br>
<br>
<code>Gene 2</code><br>
<code>max2.v9.v9.acsc.v7.v8.add3.tan.v4.v0.v0.v0.v8.C1.v1.v3.v5.v2.v8.v2.v5.v9.v2.v5.v4.v1.v2.v8.v7.v6.v0.v5.C0</code><br>
<code>C0: -9.226903090150953</code><br>
<code>C1: 0.7626090735460895</code><br>
<br>
<code>(v9-v8) + sin(min(((v8+v5)/2),v2,v9)) + max(v9,v9)</code><br>
<br>
<code>&nbsp;&nbsp;&nbsp; ...</code><br>
<br>
<code>Best Individual of Run:</code><br>
<code>Found at Generation: 4293</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 936.5758</code><br>
<code>Linking function: +</code><br>
<code>Gene 0</code><br>
<code>-.v9.v8.v0.v9.v7.acsc.v8.v2.v0.v3.v2.C1.v4.C0.v5.v9.v3.v5.v0.v2.v7.v7.v5.v9.v6.v2.v9.v2.v3.v9.v5.v5</code><br>
<code>C0: 5.233037180754705</code><br>
<code>C1: -7.143261425650365</code><br>
<br>
<code>Gene 1</code><br>
<code>sin.min3.mul3.v2.v6.cot.v6.mul4.v3.v6.v9.v8.v9.v5.v3.v0.v0.v0.v9.v5.v6.v0.v8.v4.v2.v8.v5.v4.v1.v0.C1.v5.v5</code><br>
<code>C0: 3.8794116272712955</code><br>
<code>C1: 7.460702927280668</code><br>
<br>
<code>Gene 2</code><br>
<code>v9.*.v8.v9.v8.v2.v7.*.v6.v4.v2.v7.v1.v9.v8.v8.v2.v6.v1.v8.v3.C0.v3.v3.v2.C0.C0.v9.v4.v2.v2.v7.v0</code><br>
<code>C0: 1.24681550366709</code><br>
<code>C1: -0.13544436934751047</code><br>
<br>
<code><b>(v9-v8) + sin(min((cot(v3)*v6*(v6*v9*v8*v9)),v2,v6)) + v9</b></code><br>
<br>
<code>Size of program: 17</code><br>
<code>Variables used(count): v2(1), v3(1), v6(3), v8(2), v9(4)</code><br>
<br>
<code>***** TRAINING data results *****</code><br>
<code>Statistics: </code><br>
<code>MSE:&nbsp; 24.239096329510474&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;RAE:&nbsp; 0.06769687388092159&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;RSE:&nbsp; 0.004585892048274829</code><br>
<code>RMSE: 4.923321676420349&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; MAE:&nbsp;
4.017510790472716&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; RRSE:
0.06771921476416298</code><br>
<br>
<br>
<code>***** TEST data results *****</code><br>
<code>Number of Calculation Errors: 0 out of 10 test sets</code><br>
<code>Statistics: </code><br>
<code>MSE:&nbsp; 16.72184789340559&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
RAE:&nbsp; 0.11520996098236828&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;RSE:&nbsp;
0.019535079218796565</code><br>
<code>RMSE: 4.089235612361508&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; MAE:&nbsp;
3.011454736524365&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; RRSE:
0.13976794775196696</code><br>
<br>
<code>Observed&nbsp;&nbsp;&nbsp; Computed</code><br>
<code>811.63994&nbsp;&nbsp;&nbsp; 808.827861961812</code><br>
<code>797.92192&nbsp;&nbsp;&nbsp; 797.2229981918084</code><br>
<code>786.41891&nbsp;&nbsp;&nbsp; 785.1324338242816</code><br>
<code>771.09691&nbsp;&nbsp;&nbsp; 774.3511732579256</code><br>
<code>763.19992&nbsp;&nbsp;&nbsp; 756.108499923999</code><br>
<br>
<code>&nbsp;&nbsp;&nbsp; ...</code><br>
<br>
<code>858.957&nbsp;&nbsp;&nbsp; 853.8591859872175</code><br>
<code>869.964&nbsp;&nbsp;&nbsp; 871.6444114907323</code><br>
<code>878.539&nbsp;&nbsp;&nbsp; 880.63390351821</code><br>
<br>
<code>890.288&nbsp;&nbsp;&nbsp; 886.6045236016611</code><br>
<code>900.373&nbsp;&nbsp;&nbsp; 901.3867552973694</code><br>
<code>916.525&nbsp;&nbsp;&nbsp; 910.849758599105</code><br>
<code>932.277&nbsp;&nbsp;&nbsp; 932.416393052961</code><br>
<code>947.579&nbsp;&nbsp;&nbsp; 947.2086536125466</code><br>
<code>960.562&nbsp;&nbsp;&nbsp; 962.9643013882095</code><br>
<code>966.205&nbsp;&nbsp;&nbsp; 973.9101611582124</code><br>
<code>970.634&nbsp;&nbsp;&nbsp; 970.9536648254193</code><br>
<code>972.626&nbsp;&nbsp;&nbsp; 974.1782314966911</code><br>
<code>968.324&nbsp;&nbsp;&nbsp; 975.5769759596939</code></span><br>
<br>
<b>2. Exchange Rate Example</b> - This example follows the exchange rate of the
Australian dollar against the US dollar. The original time series data file
USEXCHM.DAT can be obtained from:<br>
<br>
&nbsp;&nbsp;&nbsp; Hyndman, R.J. (n.d.) Time Series Data Library,
http://www.robhyndman.info/TSDL/. <br>
&nbsp;&nbsp;&nbsp; <span class=GramE>Accessed on June 3, 2006.</span><br>
<br>
For the integral time series used in this run, each value from the last 100
records of&nbsp;the original raw series was replaced by its moving average
using a smoothing period p = 12.<br>
<br>
<b>3. Earthquake Data Example</b> - This example follows the number of
earthquakes per year of magnitude 7.0 or greater from 1900 to 1998. The source
is the US National Earthquake Information Center and the original time series
data file EARTHQ.DAT can be obtained from:<br>
<br>
&nbsp;&nbsp;&nbsp; Hyndman, R.J. (n.d.) Time Series Data Library,
http://www.robhyndman.info/TSDL/. <span class=GramE>Accessed on June 3, 2006.</span><br>
<br>
For the integral time series used in this run, each value from original raw
series was&nbsp;replaced by its moving average using a smoothing period p = 10.<br>
<br>
<b>4. Sunspost Data Example</b> - This example uses the Wolfer sunspots data
that can be found at http://mason.gmu.edu/~csutton/sun554.html or do a search
on the internet. For this example we have 3 variations to show how the
timeseries files can be accessed. The Sunspots.param variation gets the data
from a timeseries text file, Sunspots.txt. Sunspots2.param gets the data from a
user program. Sunspots3.param get the data from a text file but it is
preprocessed already, so one should not specify the delay, or other 2
parameters that are required when the data must be processed.</p>

<p class=MsoNormal style='margin-bottom:12.0pt'><o:p>&nbsp;</o:p></p>

<h1><a name="Logical_Problem_Examples"></a>Logical Problem Examples</h1>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<p class=MsoNormal>Logical problems have data values that are all&nbsp;logical,
i.e. true or false (represented by 1 and 0 in gep). They can use only logical
functions such as <b>and</b>, <b>or</b>, <b>not</b>, etc. and the linking
function must be a logical function as well (from the list: <b>and</b>, <b>or</b>,
<b>nand</b>, <b>nor</b>, <b>xor</b>, <b>nxor</b>).&nbsp;<br>
<br>
<b>1. XOR example</b> - Given a set of x, y and result values (where the result
is the <b>exclusive or</b> of the x and y values) we try to discover an
equivalent function using only the functions <b>and</b>, <b>or</b> and <b>not</b>.
The Xor.param file contains:</p>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
<code><span style='font-size:10.0pt'>gep.species.numgenes = 2</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>gep.species.gene-headsize = 8</code><br>
<code>gep.species.problemtype = logical</code><br>
<code>gep.species.symbolset.terminalfilename = ec/gep/test/Xor.txt</code><br>
<code>gep.species.symbolset.functionsize = 3</code><br>
<code>gep.species.symbolset.function.0 = Not</code><br>
<code>gep.species.symbolset.function.0.weight = 1</code><br>
<code>gep.species.symbolset.function.1 = <span class=GramE>And</span></code><br>
<code>gep.species.symbolset.function.1.weight = 1</code><br>
<code>gep.species.symbolset.function.2 = Or</code><br>
<code>gep.species.symbolset.function.2.weight = 1</code></span><span
style='font-size:10.0pt'><o:p></o:p></span></p>

<p class=MsoNormal>The resulting statistic file shows that an expression was
discovered that matches the requirement exactly.<br>
<br>
<code><span style='font-size:10.0pt'>Seed(s) used in this job: 4357</span></code><span
style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>Problem Type: logical</code><br>
<br>
<code>Generation: 0</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 500.0</code><br>
<code>Linking function: and</code><br>
<code>Gene 0</code><br>
<code>not.and.d1.and.d1.d0.not.d0.d0.d0.d0.d1.d1.d0.d0.d1.d1</code><br>
<code>Gene 1</code><br>
<code>not.and.d0.d1.d0.d1.d0.d1.d0.d0.d0.d0.d1.d0.d1.d0.d0</code><br>
<br>
<code>(not (d1 and (d1 and d0))) and (not (d0 and d1))</code><br>
<br>
<code>Generation: 2</code><br>
<code>Best Individual:</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 1000.0</code><br>
<code>Linking function: and</code><br>
<code>Gene 0</code><br>
<code>not.and.d0.d1.d0.and.d1.not.d0.d1.d1.d0.d0.d1.d1.d0.d0</code><br>
<code>Gene 1</code><br>
<code>or.d0.or.d1.d1.not.not.not.d1.d0.d0.d1.d0.d0.d1.d0.d0</code><br>
<br>
<code>(not (d0 and d1)) and (d0 or (d1 or d1))</code><br>
<br>
<code>Best Individual of Run:</code><br>
<code>Found at Generation: 2</code><br>
<code>Evaluated: T</code><br>
<code>Fitness: 1000.0</code><br>
<code>Linking function: and</code><br>
<code>Gene 0</code><br>
<code>not.and.d0.d1.d0.and.d1.not.d0.d1.d1.d0.d0.d1.d1.d0.d0</code><br>
<code>Gene 1</code><br>
<code>or.d0.or.d1.d1.not.not.not.d1.d0.d0.d1.d0.d0.d1.d0.d0</code><br>
<br>
<code>(not (d0 and d1)) and (d0 or (d1 or d1))</code><br>
<br>
<code>Size of program: 9</code><br>
<code>Variables used(count): d0(2), d1(3)</code><br>
<br>
<code>***** TRAINING data results *****</code><br>
<code>Confusion Matrix: </code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Predicted Value</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|&nbsp; Yes&nbsp;&nbsp;&nbsp;|&nbsp; No</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|----------------</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
Yes|&nbsp; 2&nbsp;&nbsp;&nbsp; &nbsp;|&nbsp; 0</code><br>
<code>Actual Value&nbsp;&nbsp; |----------------</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; No
|&nbsp; 0&nbsp;&nbsp;&nbsp; &nbsp;|&nbsp; 2</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
|----------------</code><br>
<br>
<code>Observed&nbsp;&nbsp;&nbsp; Computed</code><br>
<code>0.0&nbsp;&nbsp;&nbsp; &nbsp; &nbsp; &nbsp;0.0</code><br>
<code>1.0&nbsp;&nbsp;&nbsp; &nbsp; &nbsp; &nbsp;1.0</code><br>
<code>1.0&nbsp;&nbsp;&nbsp; &nbsp; &nbsp; &nbsp;1.0</code><br>
<code>0.0&nbsp;&nbsp; &nbsp; &nbsp; &nbsp; 0.0</code></span><br>
<br>
<b>2. Majority Function Example</b> - This example looks for a logical
expression that can determine if given three inputs (x, y and z) there are at
least 2 true (1) values. The solution uses only the operators <b>not</b>, <b>and</b>
and <b>or</b>.<br>
<br>
<b>3. Odd 3-Parity Example</b> - This example looks at 3 <span class=GramE>boolean</span>
input variables (x, y, and z) and looks for a logical expression that can
decide if there are an odd&nbsp;number of true (1) values. The solution uses
only the operators <b>not</b>, <b>and</b> and <b>or</b>.<br>
<br>
<b>4. 6 Bit&nbsp;Multiplexer Example</b> - This example is an attempt to
discover a 6 bit multiplexer function using <b>and</b>, <b>or</b> and <b>not</b>
operators. It looks at 6 <span class=GramE>boolean</span> variables
representing 2 control bits and&nbsp;4 data bits. The 1st 2 bits are the
control bits and they identify which&nbsp;of the next 4 input bits will be in
the output. For example, if the&nbsp;control bits are 00 then the 3rd bit value
is value of the output;&nbsp;if the control bits are 01 then the 4th bit value
is the output value;&nbsp;if the control bits are 10 then the 5th bit value is
the output value;&nbsp;if the control bits are 11 then the 6th bit value is the
output value.&nbsp;This can be expressed as:<br>
<br>
&nbsp;&nbsp;&nbsp; A&amp;~c1&amp;~c2 v B&amp;~c1&amp;c2 v C&amp;c1&amp;~c2 v
D&amp;c1&amp;c2 = output<br>
<br>
&nbsp;&nbsp;&nbsp; where c1 is control bit 1 and c2 is control bit 2<br>
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp; A, B, C, D are
the 4 input bits<br>
<br>
So if we have 6 bits labelled d0, d1, d2, d3, d4, d5 and d0=c1, d1=c2, A=d2,
B=d3, C=d4, D=d5 then an excellent solution to the problem will be:<br>
<br>
&nbsp;&nbsp;&nbsp; output = d2&amp;~d0&amp;~d1 v d3&amp;~d0&amp;d1 v
d4&amp;d0&amp;~d1 v d5&amp;d0&amp;d1</p>

<p class=MsoNormal><o:p>&nbsp;</o:p></p>

<h1><a name="A_Detailed_Look_at_the_GEP_Parameter"></a>A Detailed Look at the
GEP Parameter File (gep.params)</h1>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
The file ec/gep/gep.params has most of the required details in comments. But
here we list the parameter settings that are of concern for gep problems.<br>
<br>
<b>eval.problem</b> - The user code to support the problem. <span class=GramE>For
example, ec.app.gep.test2.test2.</span> These classes will be of type
GEPProblem.&nbsp;If the user wants to just take a default route and not have to
create any code at all he can use the default GEPProblem class,
ec.gep.GEPDefaultUserProb. Then he will need to also specify a <st1:PersonName
w:st="on">'</st1:PersonName>fitness-function<st1:PersonName w:st="on">'</st1:PersonName>&nbsp;to
use by setting the parameter <b>eval.problem.fitness-function </b>and possibly
one or more parameters of the form <b>eval.problem.fitness-function-arg0</b>, <b>eval.problem.fitness-function-arg1</b>.&nbsp;<br>
<br>
<span class=GramE><b>evalthreads</b></span> and <b>breedthreads</b> &nbsp;-
should both be 1 always. It<st1:PersonName w:st="on">'</st1:PersonName>s
possible to have multiple threads but it would NOT work as per the
GeneXProTools system and I expect the code as is now will not work (at tleast
the code would need to be modified so an adjustment is made so that the
multiple paths through breeding &nbsp;would each reduce the rates for
inversion, mutation, etc.)<br>
<br>
<span class=GramE><b>checkpoint</b></span> - not well tested but it should work
OK. If not let me know. <span class=GramE>May not be important since rerunning
a problem may be fast enough.</span><br>
<span class=GramE><b>state</b></span> - Should be
ec.simple.SimpleEvolutionState; <b>do not change this!</b><br>
<span class=GramE><b>init</b></span> - Should be ec.gep.GEPInitializer;
probably should not need to change this.<br>
<span class=GramE><b>finish</b></span> - Should be ec.simple.SimpleFinisher;
probably should not need to change this.<br>
<span class=GramE><b>exch</b></span> - Should be ec.simple.SimpleExchanger;
probably should not need to change this.<br>
<span class=GramE><b>breed</b></span> -Should be ec.simple.SimpleBreeder;<b> do
not change this.<br>
<span class=GramE>eval</span></b> - Should be ec.simple.SimpleEvaluator; do not
change this without some serious thought.<br>
<span class=GramE><b>generations</b></span> - As for standard ecj, maximum
number of generations to run<br>
<b>quit-on-run-complete</b> - As for standard ecj, true if &nbsp;will quit when
an <st1:PersonName w:st="on">'</st1:PersonName>ideal<st1:PersonName w:st="on">'</st1:PersonName>
individual is found in a generation; false if will run until all generations
are completed. <span class=GramE>an</span> <st1:PersonName w:st="on">'</st1:PersonName>ideal<st1:PersonName
w:st="on">'</st1:PersonName> individual is normally when the maximum fitness
value (or something very vlose to maximum) is found. This can be controlled by
the user written code (see eval.problem) or by modifying the default user code.<br>
<b style='mso-bidi-font-weight:normal'>fitness-function-ideal-percent-threshold</b>
- specifies a percentage of the maximum fitness value (MaxThreshold) that must
be reached to be considered an ideal fitness value; the default value is
99.999999; so if the maximum fitness value is 1000 and the fitness-function-ideal-percent-threshold
is set to 99.9 then a fitness value &gt;= 999 will be considered ideal; by
default a value &gt;= 999.99999 will be ideal.</p>

<p class=MsoNormal><br>
<span class=GramE><b>pop</b></span> - Should be ec.Population. - As for
standard ecj.<br>
<b>pop.subpops</b> - must be set to 1 (<b>do not change</b> since only 1 subpop
makes sense).<br>
<b>pop.subpop.0</b> - must be ec.Subpopulation. <b>Do not change</b>.<br>
<span class=GramE><b>pop.subpop.0.duplicate-retries</b></span> - should be 0? <span
class=GramE>Haven<st1:PersonName w:st="on">'</st1:PersonName>t looked closely
at this.</span><br>
<b>pop.subpop.0.size</b> - application dependent; number of individuals in each
generation<br>
<br>
<b>pop.subpop.0.species</b> - must be ec.gep.GEPSpecies.<br>
<b>pop.subpop.0.species.fitness</b> - Should be ec.simple.SimpleFitness. <span
class=GramE>Could change this as required (for example to support a more
complex fitness or to use doubles rather than floats).</span><br>
<b>pop.subpop.0.species.ind</b> - must be ec.gep.GEPIndividual.<br>
<br>
The following are the standard defaults for the evolutionary parameters as
suggested by Ferreira. These can be changed as required. I called these
probabilities but they are actually <st1:PersonName w:st="on">'</st1:PersonName>rates<st1:PersonName
w:st="on">'</st1:PersonName> the way Ferreira uses them. If they were
probabilities as I first thought then the number of affected individuals in
each generation would vary according to the probability. But instead these
rates are used to determine the number of individuals that will be affected in
every generation (constant). The individuals are randomly chosen.<br>
<span class=GramE><b>gep.species.inversion-prob</b></span>&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; &nbsp; &nbsp; = 0.1<br>
<b>gep.species.mutation-prob&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; </b>&nbsp; &nbsp;&nbsp; = 0.044<br>
<b>gep.species.istransposition-prob</b>&nbsp;&nbsp; &nbsp; &nbsp; = 0.1<br>
<b>gep.species.ristransposition-prob&nbsp; &nbsp; &nbsp; </b>= 0.1<br>
<b>gep.species.onepointrecomb-prob&nbsp; &nbsp;&nbsp; </b>= 0.3<br>
<b>gep.species.twopointrecomb-prob&nbsp; &nbsp;&nbsp; </b>= 0.3<br>
<b>gep.species.generecomb-prob&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;</b>= 0.1<br>
<b>gep.species.genetransposition-prob</b> &nbsp;= 0.1<br>
<br>
Constants can not be used with Logical problem types and will be ignored if
specified.<br>
<span class=GramE><b>gep.species.use-constants</b></span> - set to true if
using constants else&nbsp;false.<br>
<b>gep.species.numconstantspergene</b> - Number of constants generated for each
gene.<br>
<span class=GramE><b>gep.species.integer-constants</b></span> - Set to true if
using integer constants or false to use floating point constants.<br>
<span class=GramE><b>gep.species.constants-lowerlimit</b></span> - Minimum
value of the constants chosen.<br>
<span class=GramE><b>gep.species.constants-upperlimit</b></span> - Maximum
value of the constants chosen.<br>
These are default to these values as suggested by Ferreira. They can be changed
as required for the problem, but will need to understand what they mean (read
the information in Ferreira<st1:PersonName w:st="on">'</st1:PersonName>s book,
etc.)<br>
<span class=GramE><b>gep.species.rnc-mutation-prob</b></span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
= 0.01<br>
<b>gep.species.dc-mutation-prob</b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
= 0.044<br>
<b>gep.species.dc-inversion-prob</b>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
= 0.1<br>
<b>gep.species.dc-istransposition-prob</b>&nbsp;&nbsp; = 0.1<br>
<br>
<b>gep.species.numchromosomes</b> - The number of chromosomes to be used for
the problem. This is not part of Ferreira<st1:PersonName w:st="on">'</st1:PersonName>s
implementation of GEP and is set to 1 normally when using the standard GEP of
Ferreira. It is a new feature added at NRC to deal with multi criteria decision
making problems (represented by multiple dependent variables in the data set).
When used, it will require a special fitness function to deal with the multi
chromosomes: like the SEWD (Sammon Error with Dissimilarity) fitness function.<br>
<b>gep.species.numgenes</b> - The number of genes to be used for the problem.<br>
<span class=GramE><b>gep.species.gene-headsize</b> - The length of the head of
each gene.</span><br>
<br>
The <b style='mso-bidi-font-weight:normal'><i style='mso-bidi-font-style:normal'>linking-function</i></b>
is the function to use to combine the values of the expressions for each gene.
In early versions the linking-function could be one of +, -, * or / for
numerical problem types and for&nbsp;logic problems it could be one of: and,
or, nand, nor, xor, nxor. This was as provided in Ferreira’s implementation.
These choices are still available, but now any function defined as a
GEPFunctionSymbol (such as <span style='font-size:11.0pt;font-family:"Courier New"'>Add,
Sub, Mul, Div, Ifgt4, Ifeq4, <span class=GramE>Pow</span>,</span> etc) that has
arity of 2 or more can be used. It must be specified with the same ‘case’ as
the class name and there must be a ‘correct’ number of genes that suit the
arity of the function. For a function with arity of 2 any number of genes is
OK; for arity 3 there must be 3, 5, 7, 9, etc. genes; for arity 4 there must be
4, 7, 10, 13, etc. genes; For 1 gene then no function is needed.<br>
<span class=GramE><b>gep.species.gene-linking-function</b></span> - The
function to use when combining the values the expressions for each gene.<br>
<br>
Problem type must be one of: functionfinding, classification, <span
class=GramE>timeseries</span>, logical.&nbsp;Set default to be <st1:PersonName
w:st="on">'</st1:PersonName>unknown<st1:PersonName w:st="on">'</st1:PersonName>
so user is forced to specify in the problem params file.<br>
<b>gep.species.problemtype</b> = unknown<br>
<br>
<b>gep.species.classification-threshold</b> - If the problem is a <b>classification</b>
type problem then a threshold value (used&nbsp;to convert real values to 0 or 1
during fitness calculations) must be&nbsp;specified.<br>
<br>
If the problem type is timeseries then the user needs to specify a number of
parameters. See ec.gep.gep.params for full details.<br>
<span class=GramE><b>gep.species.timeseries-delay</b></span><b>&nbsp;<br>
gep.species.timeseries-embeddingdimension<br>
gep.species.timeseries-testingpredictions</b>&nbsp;<br>
<br>
<b>gep.species.symbolset</b>&nbsp; - Must be ec.gep.GEPSymbolSet; <b>do NOT
change</b>.<br>
<b>gep.species.symbolset.name</b> - can be any string but must be specified;
not actually used (at one point it seemed we might have multiple symbols sets
but this will not likely happen0; should be removed at some point.<br>
<br>
<span class=GramE>Examples for specifiying the functions to be used in the
problem and their weights.</span> The set of functions can be found by looking
at the directory ec/.gep/symbols.<br>
<b>gep.species.symbolset.functionsize</b> = 4<br>
<b>gep.species.symbolset.function.0</b> = Add<br>
<b>gep.species.symbolset.function.0.weight</b> = 1<br>
<b>gep.species.symbolset.function.1</b> = Sub<br>
<b>gep.species.symbolset.function.1.weight</b> = 1<br>
<b>gep.species.symbolset.function.2</b> = Mul<br>
<b>gep.species.symbolset.function.2.weight</b> = 1<br>
<b>gep.species.symbolset.function.3</b> = Div<br>
<b>gep.species.symbolset.function.3.weight</b> = 1<br>
<br>
The number and names of the terminal symbols (independent variables) can be
specified as shown below:<br>
<b>gep.species.symbolset.terminalsize </b>= 1<br>
<b>gep.species.symbolset.terminal.0</b> = x<br>
Or for the terminal set, the info can be in a text file where the file&nbsp;has
the names of the terminal symbols and the dependent variable in the first
record/line (comma separated, space or tab separated, etc)&nbsp;and the
training values for the terminal symbols in the other&nbsp;records; the number
of terminals is computed from the number of terminal symbols in the 1st record.
See details in ec/gep/gep.params and look at some examples.<br>
<b>gep.species.symbolset.terminalfilename</b> = filename<br>
<b>gep.species.symbolset.terminalfileseparator</b> <span class=GramE>= ,</span><br>
or for a tab<br>
<b>gep.species.symbolset.terminalfileseparator</b> = tab<br>
or for a space<br>
<b>gep.species.symbolset.terminalfileseparator</b> = space<br>
<br>
If there is testing data as well it can be specified. Again see full details in
ec/gep/gep.params since there are some slightly complex situations (timeseries
data etc.).<br>
<b>gep.species.symbolset.testingterminalfilename</b> = testDataFileName<br>
<br>
<span class=GramE>For</span> gep systems breeding should be as shown below. Changing
theseis possible but will not be as per original gep by Ferreira (as in
GeneXProTools) and will require some in-depth knowledge!<br>
<b>gep.species.pipe</b> = ec.gep.breed.DcGeneticOperatorsPipeline<br>
<b>gep.dcgeneticoperators.source.0</b> = ec.gep.breed.GenerecombinationPipeline<br>
<b>gep.generecombination.source.0</b> =
ec.gep.breed.TwopointrecombinationPipeline<br>
<b>gep.twopointrecombination.source.0</b> =
ec.gep.breed.OnepointrecombinationPipeline<br>
<b>gep.onepointrecombination.source.0</b> =
ec.gep.breed.GenetranspositionPipeline<br>
<b>gep.genetransposition.source.0</b> = ec.gep.breed.RIStranspositionPipeline<br>
<b>gep.RIStransposition.source.0</b> = ec.gep.breed.IStranspositionPipeline<br>
<b>gep.IStransposition.source.0</b> = ec.gep.breed.InversionPipeline<br>
<b>gep.inversion.source.0</b> = ec.gep.breed.MutationPipeline<br>
# FitProportionateSelection (roulette selection) is used exclusively by
Ferreira - could be changed<b><br>
gep.mutation.source.0 = </b>ec.select.FitProportionateSelection<b><br>
</b><br>
<b>breed.elite.0 </b>- Should be set to 1 since this is the only option used by
Ferreira; however you might want to experiment with this to see if it improves
the evolution for your problem or just to force the best n individuals to be
carried forward.<br>
<br>
We supply a not so simple GEPSimpleStatistics class to handle the basic capture
of information from a gep run. For the basic setting of statistics output you
can do something like the following.<br>
<br>
<span class=GramE><b>stat</b></span> = ec.gep.GEPSimpleStatistics<br>
<b>stat.file</b> = ourProblem.stat<br>
<b>stat.file-observed-versus-computed</b> = ourOvsC.stat<br>
<b>stat.file-observed-versus-computed-test</b> = ourOvsCtest.stat<br>
<b>stat.detail-to-log</b> = change<br>
<b>stat.number-of-best-to-log</b> = 1<br>
<br>
This contains all of the parameters that you might like to use but variations
can be done depending what you want to capture. As shown above,
GEPSimpleStatistics has a number of parameters it&nbsp;can use to control the
output<span class=GramE>:</span><br>
<b>stat.file </b>- The main output is directed to the file specified by the
file parameter.<br>
<span class=GramE><b>file-observed-versus-computed</b></span> - this file, if
specified, gets the results&nbsp;of the best of run model by listing the
observed values versus&nbsp;the model<st1:PersonName w:st="on">'</st1:PersonName>s
predicted values (useful for plotting). By default this&nbsp;will include the
training and testing data values (with test values&nbsp;appended to the end if
they exist). To get the training and testing&nbsp;values separated identify a
second file in the parameter<span class=GramE>:</span><br>
<b>file-observed-versus-computed-test</b>. If <b>file-observed-versus-computed</b>
is<b> </b>not specified it prints to the&nbsp;main output file, if specified,
or to the console.<br>
<span class=GramE><b>file-observed-versus-computed-test</b></span> - If
specified, this identifies a&nbsp;file with only the computed test data values.
If not specified it prints to the console.<br>
<span class=GramE><b>no-observed-versus-computed</b></span> - If <b>true</b>
then don<st1:PersonName w:st="on">'</st1:PersonName>t print any
observed/computed values&nbsp;even if files are specified. Default is <b>false</b>.<br>
<span class=GramE><b>detail-to-log</b></span> - the user can specify the detail
of the output to the main log file (param <st1:PersonName w:st="on">'</st1:PersonName>stat.file<st1:PersonName
w:st="on">'</st1:PersonName>). Values can be one of:<br>
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;&nbsp;<b>all</b> - display the best model
from every generation (can be a large&nbsp;amount of output)<br>
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; <b>change</b> - logs the best model of
the generation when the fitness improves (default)<br>
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; <b>final</b> - logs only the best of run
model information<br>
<b>number-of-best-to-log</b> - this determines the number of best individuals
to display in the final results. By default it is 1.<br>
<br>
For the multiple chromosome case there is another not so simple
GEPSimpleStatisticsMultipleChromosomes class to handle the capture of
information from a gep run. For the basic setting of statistics output you can
do something like the following.<br>
<br>
<span class=GramE><b>stat</b></span> =
ec.gep.GEPSimpleStatisticsMultipleChromosomes<br>
<b>stat.file</b> = ourProblem.stat<br>
<b>stat.file-computed-values</b> = ourOvsC.stat<br>
<b>stat.no-computed</b> = false<br>
<b>stat.detail-to-log</b> = change<br>
<b>stat.number-of-best-to-log</b> = 1<br>
<br>
These are similar to the parameters in the GEPSimpleStatistics case but we don<st1:PersonName
w:st="on">'</st1:PersonName>t display the observed/computed pairs (just the
computed values for the multiple dependent variables) so we only need the <b>stat.file-computed-values
</b>and<b> stat.no-computed </b>parameters. At some point we should (will?) get
the parameters to be the same so we don<st1:PersonName w:st="on">'</st1:PersonName>t
need different ones for the 2 cases.<br>
<br>
More complex things can be done as per standard ECJ. Consider the example below
examples for setting multiple statistic packages.<br>
<span class=GramE>stat.num-children</span> = 3<br>
stat.child.0 = ec.gep.test.SimpleXYSeriesChartStatistics<br>
stat.child.0.title = Best of Generation<br>
stat.child.0.x-axis-label = generation<br>
stat.child.0.y-axis-label = fitness<br>
stat.child.1 = ec.simple.SimpleShortStatistics<br>
stat.child.1.file = simpleshortstats.txt<br>
stat.child.1.gather-full = true<br>
stat.child.2 = ec.gep.GEPSimpleStatistics<br>
stat.child.2.file = simplestats.txt<br>
stat.child.2.file-observed-versus-computed = simplestatsOvsC.txt</p>

<p class=MsoNormal style='margin-bottom:12.0pt'><br>
Note that as revisions are made, new parameters may be added or changed, so it
is best to also look at the gep.params file for details.</p>

<h1><a name="Some_of_the_Implementation_Details"></a>Some of the Implementation
Details<br>
<br>
<span class=GramE><span style='font-weight:normal'>The</span></span><span
style='font-weight:normal'> system has been tested using </span>Java version
1.6<span style='font-weight:normal'>.</span><br>
<br>
<span style='font-weight:normal'>One thing to note about the gep implementation
is that </span>all values are double<span style='font-weight:normal'>, rather
than float as in other ECJ parts. The only thing not in double is the fitness
and this only because we used the SimpleFitness provided with ecj. When
complicated expressions are created during the evolution, considerable loss of
precision creeps in when they are evaluated as floats rather than doubles. The
penalty may be a loss of speed but I don<st1:PersonName w:st="on">'</st1:PersonName>t
think it<st1:PersonName w:st="on">'</st1:PersonName>s too great. The gep
implementation is still quite speedy and we<st1:PersonName w:st="on">'</st1:PersonName>ve
done a few demanding examples (50 individuals in the population, 5 genes, head
size of 15, 2 constants per gene, 13 functions, 25000 generations) and the
performance is quite acceptable. Try some of the examples or your own problems
and I think you<st1:PersonName w:st="on">'</st1:PersonName>ll see that the
performance is quite good. Note also that the logical problems, although they
use only values 0 and 1 are still done with double values. This could be much
improved I suppose,&nbsp;but no effort has been made to do so. It works fairly
quickly as is and these are not problems we were all that interested in for our
research.<br>
<br>
This is the first version of this gep extension and the code could use some
reorganization, optimization and perhaps better adherence to the ECJ structures
(although it isn<st1:PersonName w:st="on">'</st1:PersonName>t too bad).<br>
<br>
<a name=ModsToExistingECJcode></a>I did modify a couple of the standard ECJ
files. These were very minor modifications. First&nbsp;the </span>ec.display.Console<span
style='font-weight:normal'> was changed slightly to allow the seed that is used
in a run to be stored for access later in statistics output. This also required
some changes to other files (below). The change was near line 935 in
ec.display.Console:<br>
<br>
</span><code><span style='font-size:10.0pt;font-weight:normal'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
state.evalthreads = evalthreads;</span></code><span style='font-size:10.0pt;
font-family:"Courier New";font-weight:normal'><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;state.breedthreads =
breedthreads;</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;state.seeds = seeds<span
class=GramE>; &nbsp;/</span>/ this line was added</code></span><br>
<br>
<span style='font-weight:normal'>Corresponding changes were made to files </span>ec.Evolve<span
style='font-weight:normal'> and </span>ec.EvolutionState <span
style='font-weight:normal'>to support the need to have access to the seed
value(s) for reporting in statistics. This is important so that one can rerun
the program with the same seed at some later time and there was no way to
record this in the standard version of ECJ. In ec.Evolve near line 371 we added
the code:<br>
<br>
</span><code><span style='font-size:10.0pt;font-weight:normal'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
state.breedthreads = breedthreads;</span></code><span style='font-size:10.0pt;
font-family:"Courier New";font-weight:normal'><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; state.randomSeedOffset =
randomSeedOffset;</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; state.seeds = seeds<span
class=GramE>; &nbsp;/</span>/ this line was added</code><br>
</span><span style='font-weight:normal'><br>
and near line 187 in ec.EvolutionState we added the code:<br>
<br>
</span><code><span style='font-size:10.0pt;font-weight:normal'>&nbsp;&nbsp;&nbsp;
/**</span></code><span style='font-size:10.0pt;font-family:"Courier New";
font-weight:normal'><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; * The seeds used for the current job</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; */</code><br>
<code>&nbsp;&nbsp;&nbsp; public int seeds[];</code></span><span
style='font-weight:normal'><br>
<br>
<br>
<a name=AddingNewFunctions></a>The two areas where users might want to add
functionality to this gep implementation are </span>new mathematical/logical
functions<span style='font-weight:normal'> and </span>new fitness functions<span
style='font-weight:normal'>. All mathematical functions are derived from the
abstract class GEPFunctionSymbol. Any new function will have to provide just a
few simple methods. Let<st1:PersonName w:st="on">'</st1:PersonName>s look at
the Add function.<br>
<br>
</span><code><span style='font-size:10.0pt;font-weight:normal'>package
ec.gep.symbols;</span></code><span style='font-size:10.0pt;font-family:"Courier New";
font-weight:normal'><br>
<code>import ec.EvolutionState;</code><br>
<code>import ec.gep.GEPFunctionSymbol;</code><br>
<code>import ec.gp.GPData;</code><br>
<code>import ec.gp.GPIndividual;</code><br>
<code>import ec.util.Parameter;</code><br>
<br>
<code>public class Add extends GEPFunctionSymbol {</code><br>
<br>
<code>&nbsp;&nbsp;&nbsp; public Add() </code><br>
<code>&nbsp;&nbsp;&nbsp; {</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; super(&quot;+&quot;, 2);</code><br>
<code>&nbsp;&nbsp;&nbsp; }</code><br>
<br>
<code>&nbsp;&nbsp;&nbsp; public double eval(double params[]) </code><br>
<code>&nbsp;&nbsp;&nbsp; {</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; //should check that there are 2
params</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; return (params[0] + params[1]);</code><br>
<code>&nbsp;&nbsp;&nbsp; }</code><br>
<code>&nbsp;&nbsp;&nbsp; </code><br>
<code>&nbsp;&nbsp;&nbsp; public boolean isLogicalFunction()</code><br>
<code>&nbsp;&nbsp;&nbsp; {</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; return false;</code><br>
<code>&nbsp;&nbsp;&nbsp; }</code><br>
<br>
<code>&nbsp;&nbsp;&nbsp; public String getMathExpressionAsString( String p[] )</code><br>
<code>&nbsp;&nbsp;&nbsp; {</code><br>
<code>&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; return &quot;(&quot; + p[0] +
&quot;+&quot; + p[1] + &quot;)&quot;;</code><br>
<code>&nbsp;&nbsp;&nbsp; }</code><br>
<code>}</code></span></h1>

<p class=MsoNormal><br>
The constructor will simply call the super constructor and pass the symbol to
be used when printing the Karva representation of the expressions formed in the
genes. It will also pass the <st1:PersonName w:st="on">'</st1:PersonName>arity<st1:PersonName
w:st="on">'</st1:PersonName> of the function (how many arguments it needs when
executed). It must also implement an <b>eval</b> method to calculate <span
class=GramE>it<st1:PersonName w:st="on">'</st1:PersonName>s</span> value using
the arguments passed to it. As the comment says some checking could (should?)
be done to ensure there are 2 arguments available in the params array, but this
is probably not the place to do it. It should be done in the place where eval
is called so every function doesn<st1:PersonName w:st="on">'</st1:PersonName>t
have to do it (i.e. in the eval method of GEPExpressionTreeNode). There are
places where it is necessary to know if this is a logical function or not (so
we can ensure that logical problems use only logical functions and other
problem types don<st1:PersonName w:st="on">'</st1:PersonName>t use logical
functions. so we must provide the simple isLogicalFunction method. Finally in
order to display the expressions in a mathematical notation it is necessary for
the function to provide an appropriate string representation for this purpose.
I<st1:PersonName w:st="on">'</st1:PersonName>d suggest looking at a similar
function to see how it was done and proceed from there. The array passed to <code><b><span
style='font-size:10.0pt'>getMathExpressionAsString</span></b></code> contains
the string representations of its parameters.<br>
<br>
Adding <b>new fitness functions</b> is done by adding some methods to the class
ec.gep.GEPFitnessFunction. For each of the fitness functions there are 3
methods that need to be supplied: one to calculate the fitness&nbsp;(with a
value from 0 to some maximum); one to give the raw fitness value, the value
prior&nbsp;to being mapped between 0 and the maximum value; and one to provide
the maximum value that the fitness value can take for that fitness function. Actually
the raw value is not always needed (see for example the specialized fitness
function WCorrRMSE (Weighted correlation coefficient and Root Mean Squared
Error). There is no raw value for this since it combines to other fitness
values. <span class=GramE>but</span> since some users might want access to the
value before it is converted to a value from 0 to a maximum value, it seemed
prudent to make this available whenever possible. By convention we use the
names XXXXfitness, XXXXrawFitness&nbsp;and XXXXmaxFitness for all of the
fitness functions. XXXX is the pneumonic for the&nbsp;fitness function (e.g.
RRSE, MAE, etc.). This allows a user to specifiy a fitness function to
be&nbsp;used in the default user program by specifying only the short pneumonic
for the fitness function (they don<st1:PersonName w:st="on">'</st1:PersonName>t
provide ANY code if they use this - check the code in GEPDefaultUserProg.java).
The XXXXmaxFitness functions&nbsp;all have 1&nbsp;arg when most don<st1:PersonName
w:st="on">'</st1:PersonName>t need one.&nbsp;This was just to make it
consistent and easier to call the method based on the XXXX part only (again
see&nbsp;GEPDefaultUserProg.java).&nbsp;For example the functions for the MSE
fitness functions are called as:<br>
<br>
<code><span style='font-size:10.0pt'>&nbsp; &nbsp;&nbsp;&nbsp; <span
class=GramE>GEPFitnessFunction.MSEfitness(</span>useTrainingData, gepindividual
);</span></code><span style='font-size:10.0pt;font-family:"Courier New"'><br>
<code>&nbsp; &nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; - calculates the fitness
using the individual<st1:PersonName w:st="on">'</st1:PersonName>s gene
expressions<o:p></o:p></code></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='mso-spacerun:yes'>    </span>- it uses either the training date (<code><span
style='font-size:10.0pt'>useTrainingData</span></code><span style='color:black'>
is true) or the testing data (</span><code><span style='font-size:10.0pt'>useTrainingData</span></code><span
style='color:black'> is false)</span><br>
<code><span style='font-size:10.0pt'>&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; - it gets the raw fitness first from MSErawFitness (the mean
squared error</span></code><br>
<code><span style='font-size:10.0pt'>&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; &nbsp; of the predicted values versus the text/expected
values)</span></code><br>
<code><span style='font-size:10.0pt'>&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; - then it normalizes the result between 0 and 1000&nbsp; </span></code><br>
<code><span style='font-size:10.0pt'>&nbsp; &nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp; (1000 * (1/(1 + raw MSE))</span></code><br>
<code><span style='font-size:10.0pt'>&nbsp; &nbsp;&nbsp;&nbsp;
GEPFitnessFunction.MSErawFitness(useTrainingData, gepindividual,
chromosomeNum);</span></code><br>
<code><span style='font-size:10.0pt'>&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - sum((predicted valuei - expected
valuei)**2)/n</span></code><br>
<code><span style='font-size:10.0pt'>&nbsp; &nbsp;&nbsp;&nbsp;
GEPFitnessFunction.NHmaxFitness( gepindividual );</span></code><br>
<code><span style='font-size:10.0pt'>&nbsp;&nbsp; &nbsp;&nbsp;
&nbsp;&nbsp;&nbsp; - in this case max is always 1000 but some fitness functions</span></code><br>
<code><span style='font-size:10.0pt'>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;
have a maximum that depends on the number of training data sets</span></code><br>
<br>
Again let<st1:PersonName w:st="on">'</st1:PersonName>s look at a the code of a
fitness function to see what<st1:PersonName w:st="on">'</st1:PersonName>s
required.<br>
<br>
<span style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span></span><span style='font-size:10.0pt;
font-family:"Courier New";color:#3F5FBF'>/**</span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>Calculates</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>the</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><st1:PersonName
w:st="on"><span style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>'</span></st1:PersonName><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>raw<st1:PersonName
w:st="on">'</st1:PersonName></span><span style='font-size:10.0pt;font-family:
"Courier New";color:black'> </span><span style='font-size:10.0pt;font-family:
"Courier New";color:#3F5FBF'>fitness</span><span style='font-size:10.0pt;
font-family:"Courier New";color:black'> </span><span style='font-size:10.0pt;
font-family:"Courier New";color:#3F5FBF'>for</span><span style='font-size:10.0pt;
font-family:"Courier New";color:black'> </span><span style='font-size:10.0pt;
font-family:"Courier New";color:#3F5FBF'>the</span><span style='font-size:10.0pt;
font-family:"Courier New";color:black'> </span><span style='font-size:10.0pt;
font-family:"Courier New";color:#3F5FBF'>RAE</span><span style='font-size:10.0pt;
font-family:"Courier New";color:black'> </span><span style='font-size:10.0pt;
font-family:"Courier New";color:#3F5FBF'>(Relative</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>Absolute</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>Error)</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>type</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
class=GramE><span style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>fitness</span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>(before</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>the</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>normalization</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>from</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>0</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>to</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>max</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>value</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>is</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>done).</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F9FBF'>@param</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>useTrainingData</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>true</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>if</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>using</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>training</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>data,</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>else</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>use</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>testing</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>data</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F9FBF'>@param</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><st1:place
w:st="on"><st1:State w:st="on"><span class=GramE><span style='font-size:10.0pt;
  font-family:"Courier New";color:#3F5FBF'>ind</span></span></st1:State></st1:place><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>the</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>GEP</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>individual</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>that</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>needs</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>its</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>fitness</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>calculated.</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F9FBF'>@param</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>chromosomeNum</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>which</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>chromosome</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>in</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>the</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>individual</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>to</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>use</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>to</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>calc</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>the</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>raw</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>RAE</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F9FBF'>@return</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>the</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><st1:PersonName
w:st="on"><span style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>'</span></st1:PersonName><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>raw<st1:PersonName
w:st="on">'</st1:PersonName></span><span style='font-size:10.0pt;font-family:
"Courier New";color:black'> </span><span style='font-size:10.0pt;font-family:
"Courier New";color:#3F5FBF'>fitness</span><span style='font-size:10.0pt;
font-family:"Courier New";color:black'> </span><span style='font-size:10.0pt;
font-family:"Courier New";color:#3F5FBF'>value</span><span style='font-size:
10.0pt;font-family:"Courier New";color:black'> </span><span style='font-size:
10.0pt;font-family:"Courier New";color:#3F5FBF'>before</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>normalization</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>from</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>0</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>to</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>max</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>value</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*/</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>public</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>static</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>double</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> <span
style='background:silver;mso-highlight:silver'>RAErawFitness</span>(</span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>boolean</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'>
useTrainingData, GEPIndividual <st1:place w:st="on"><st1:State w:st="on">ind</st1:State></st1:place>,
</span><b><span style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>int</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> chromosomeNum) </span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span>{</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>double</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'>
sumOfAbsoluteError = 0.0;</span><span style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>double</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> expectedResult;</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>double</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> result;</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>double</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> error;</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span>GEPDependentVariable dv;</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>if</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'>
(useTrainingData) </span><span style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span><span style='mso-tab-count:1'>    </span><span
class=GramE>dv</span> = GEPDependentVariable.</span><i><span style='font-size:
10.0pt;font-family:"Courier New";color:#0000C0'>trainingData</span></i><span
style='font-size:10.0pt;font-family:"Courier New";color:black'>;</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>else</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span><span style='mso-tab-count:1'>    </span><span
class=GramE>dv</span> = GEPDependentVariable.</span><i><span style='font-size:
10.0pt;font-family:"Courier New";color:#0000C0'>testingData</span></i><span
style='font-size:10.0pt;font-family:"Courier New";color:black'>;</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>double</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> dvValues[] =
dv.getDependentVariableValues(chromosomeNum);</span><span style='font-size:
10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>double</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'>
dvSumOfAbsoluteError = dv.getDependentVariableSumOfAbsoluteError(chromosomeNum);</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>for</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> (</span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>int</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> i=0;
i&lt;dvValues.</span><span style='font-size:10.0pt;font-family:"Courier New";
color:#0000C0'>length</span><span style='font-size:10.0pt;font-family:"Courier New";
color:black'>; i++)</span><span style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span>{</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>            </span><span class=GramE>expectedResult</span>
= dvValues[i];</span><span style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>            </span><span class=GramE>result</span> =
ind.eval(chromosomeNum, useTrainingData, i);</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>            </span><span class=GramE>error</span> =
result - expectedResult;</span><span style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>            </span><span class=GramE>sumOfAbsoluteError</span>
+= Math.<i>abs</i>(error); </span><span style='font-size:10.0pt;font-family:
"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span>}</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>if</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'>
(dvSumOfAbsoluteError == 0.0)</span><span style='font-size:10.0pt;font-family:
"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span>{<span style='mso-spacerun:yes'>  
</span><span class=GramE>dvSumOfAbsoluteError</span> = </span><i><span
style='font-size:10.0pt;font-family:"Courier New";color:#0000C0'>RELATIVE_ERROR_ZERO_FACTOR</span></i><span
style='font-size:10.0pt;font-family:"Courier New";color:black'>;</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span><span style='mso-tab-count:1'>    </span><span
class=GramE>System.<i><span style='color:#0000C0'>err</span></i>.println(</span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#2A00FF'>&quot;Warning:
sum of error for dependent variable is 0 in RAE fitness calculation. Adjusting
to avoid division by zero.&quot;</span><span style='font-size:10.0pt;
font-family:"Courier New";color:black'>);</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span>}</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span style='font-size:10.0pt;
font-family:"Courier New";color:#3F7F5F'>// the raw fitness ... RAE</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>return</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'>
(sumOfAbsoluteError/dvSumOfAbsoluteError);</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span>}</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>    </span><span style='mso-tab-count:1'>  </span></span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span></span><span style='font-size:10.0pt;
font-family:"Courier New";color:#3F5FBF'>/**</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>Calculates</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>the</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>fitness</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>for</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>the</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>RAE</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>(Relative</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>Absolute</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>Error)</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>type</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>fitness.</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>Gets</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>the</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>raw</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>fitness</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>and</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>then</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>normalizes</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>between</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>0</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>and</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>max</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>value</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>as</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>(maxValue</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>(1<span
class=GramE>/(</span>1+RAE)).</span><span style='font-size:10.0pt;font-family:
"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F9FBF'>@param</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>useTrainingData</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>true</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>if</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>using</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>training</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>data,</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>else</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>use</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>testing</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>data</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F9FBF'>@param</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><st1:place
w:st="on"><st1:State w:st="on"><span class=GramE><span style='font-size:10.0pt;
  font-family:"Courier New";color:#3F5FBF'>ind</span></span></st1:State></st1:place><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>the</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>GEP</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>individual</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>that</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>needs</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>its</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>fitness</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>calculated.</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F9FBF'>@return</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>the</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>fitness</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>value</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>after</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>normalization</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>from</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>0</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>to</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>max</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>value</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*/</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>      </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>public</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>static</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>double</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> RAEfitness(</span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>boolean</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'>
useTrainingData, GEPIndividual <st1:place w:st="on"><st1:State w:st="on">ind</st1:State></st1:place>)
</span><span style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>      </span>{</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>double</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> RAE = <i><span
style='background:silver;mso-highlight:silver'>RAErawFitness</span></i>(useTrainingData,
<st1:place w:st="on"><st1:State w:st="on">ind</st1:State></st1:place>, 0);</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span style='font-size:10.0pt;
font-family:"Courier New";color:#3F7F5F'>// raw fitness is normalized between 0
and <span class=GramE>1000<span style='mso-spacerun:yes'>  </span>(</span>1000
* (1/(1+RAE))</span><span style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>        </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>return</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'>
(1000.0)/(1.0+RAE);</span><span style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-spacerun:yes'>      </span>}</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span></span><span style='font-size:10.0pt;
font-family:"Courier New";color:#3F5FBF'>/**</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>The</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>max</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>value</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>for</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>this</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>type</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>of</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>fitness</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>is</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>always</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>1000.</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F9FBF'>@param</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><st1:place
w:st="on"><st1:State w:st="on"><span class=GramE><span style='font-size:10.0pt;
  font-family:"Courier New";color:#3F5FBF'>ind</span></span></st1:State></st1:place><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>the</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>GEP</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>individual</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>that</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>needs</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>its</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>fitness</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>calculated.</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F9FBF'>@return</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>value</span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>1000.0</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span><span style='mso-spacerun:yes'> </span></span><span
style='font-size:10.0pt;font-family:"Courier New";color:#3F5FBF'>*/</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>public</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>static</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> </span><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>double</span></b><span
style='font-size:10.0pt;font-family:"Courier New";color:black'>
RAEmaxfitness(GEPIndividual <st1:place w:st="on"><st1:State w:st="on">ind</st1:State></st1:place>)
</span><span style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span>{</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:2'>            </span></span><span style='font-size:10.0pt;
font-family:"Courier New";color:#3F7F5F'>// always 1000</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:2'>            </span></span><span class=GramE><b><span
style='font-size:10.0pt;font-family:"Courier New";color:#7F0055'>return</span></b></span><span
style='font-size:10.0pt;font-family:"Courier New";color:black'> 1000.0;</span><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New";color:black'><span
style='mso-tab-count:1'>      </span>}</span><span style='font-size:10.0pt;
font-family:"Courier New"'><o:p></o:p></span></p>

<p class=MsoNormal style='mso-layout-grid-align:none;text-autospace:none'><span
style='font-size:10.0pt;font-family:"Courier New"'><o:p>&nbsp;</o:p></span></p>

<p class=MsoNormal style='text-indent:.5in'><span
style='mso-spacerun:yes'> </span>The code is fairly straight forward. the only thing
to note is that at one point we accessed the sum of the absolute error of the
values of the dependent variable data in the training set using -
GEPDependentVariable.getDependentVariableSumOfAbsoluteError(). When the
training and testing data for the program is obtained, a number of values are
calculated and stored statically in the class GEPDependentVariable. This is so
these values, which are used by several fitness functions, don<st1:PersonName
w:st="on">'</st1:PersonName>t have to be recalculated many times as the
evolutionary process measures the fitness of many individuals. The values
available for use in the fitness functions are:</p>

<p class=MsoNormal style='text-indent:.5in'><o:p>&nbsp;</o:p></p>

<p class=MsoNormal style='margin-bottom:12.0pt'><code><span style='font-size:
10.0pt'>&nbsp;&nbsp;&nbsp; /** Mean value of the training values for the
dependent variable */</span></code><span style='font-size:10.0pt;font-family:
"Courier New"'><br>
<code>&nbsp;&nbsp;&nbsp; static double dependentVariableMean = 0.0;</code><br>
<code>&nbsp;&nbsp;&nbsp; /** Sum of the absolute error between dependent
variables training values and their mean value. The</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; *<span class=GramE>&nbsp; absolute</span> error
is Abs(dv[i] - dvMean).</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; */</code><br>
<code>&nbsp;&nbsp;&nbsp; static double dependentVariableSumOfAbsoluteError =
0.0;</code><br>
<code>&nbsp;&nbsp;&nbsp; /** Sum of the squares of the absolute errors between
dependent variables training values and their mean </code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; *<span class=GramE>&nbsp; value</span>. The
absolute error is <span class=GramE>Abs(</span>dv[i] - dvMean).</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; */</code><br>
<code>&nbsp;&nbsp;&nbsp; static double
dependentVariableSumOfSquaredAbsoluteError = 0.0;</code><br>
<code>&nbsp;&nbsp;&nbsp; /** Sum of the relative errors between dependent
variables training values and their mean value.</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; *<span class=GramE>&nbsp; The</span> relative
error is Abs((dv[i] - dvMean)/dvMean).</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; */</code><br>
<code>&nbsp;&nbsp;&nbsp; static double dependentVariableSumOfRelativeError =
0.0;</code><br>
<code>&nbsp;&nbsp;&nbsp; /** Sum of the squares relative errors between
dependent variables training values and their mean value.</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; *<span class=GramE>&nbsp; The</span> relative
error is Abs((dv[i] - dvMean)/dvMean).</code><br>
<code>&nbsp;&nbsp;&nbsp;&nbsp; */</code><br>
<code>&nbsp;&nbsp;&nbsp; static double
dependentVariableSumOfSquaredRelativeError = 0.0;</code></span><span
style='font-size:10.0pt'><o:p></o:p></span></p>

<p class=MsoNormal>These values are available for both the training data as
shown above and for the test data as well. See the class ec.gep.GEPDependentVariable.</p>

<p class=MsoNormal style='margin-bottom:12.0pt'><o:p>&nbsp;</o:p></p>

</div>

</div>

</body>

</html>
